

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>8. R session &mdash; Physalia Paleogenomics 0.1.0 documentation</title>
  

  
  
  
  

  
  <script type="text/javascript" src="_static/js/modernizr.min.js"></script>
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script type="text/javascript" src="_static/jquery.js"></script>
        <script type="text/javascript" src="_static/underscore.js"></script>
        <script type="text/javascript" src="_static/doctools.js"></script>
        <script type="text/javascript" src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="9. Material from previous courses" href="11_old_material.html" />
    <link rel="prev" title="7. DO-IT-YOURSELF" href="9_DIY.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home"> Physalia Paleogenomics
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="1_ListTools.html">1. List of Tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="2_ReadsFiltering.html">2. Quality filtering of reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="3_Metagenomics_v2.html">3. Metagenomic screening of shotgun data</a></li>
<li class="toctree-l1"><a class="reference internal" href="4_ReadsMapping_v2.html">4. Alignment of reads to a reference genome</a></li>
<li class="toctree-l1"><a class="reference internal" href="7_VariantsCall.html">5. Variant calling and visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="8_Filtering_SNPs.html">6. Filtering, annotating and combining SNPs</a></li>
<li class="toctree-l1"><a class="reference internal" href="9_DIY.html">7. DO-IT-YOURSELF</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">8. R session</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#barplot">8.1. Barplot</a></li>
<li class="toctree-l2"><a class="reference internal" href="#upgma">8.2. UPGMA</a></li>
<li class="toctree-l2"><a class="reference internal" href="#nmds">8.3. nMDS</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="11_old_material.html">9. Material from previous courses</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Physalia Paleogenomics</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>8. R session</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/10_R_session.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="r-session">
<h1>8. R session<a class="headerlink" href="#r-session" title="Permalink to this headline">Â¶</a></h1>
<p>In this hands-on we will use R to run analyses and create charts from the abundance tables that we generated with Kraken2.</p>
<p>First of all, in R, we must set up the folder which contains the abundance table:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="kp">setwd</span><span class="p">(</span><span class="s">&quot;~/Documents/WORK/2-PROJECTS/HiddenFoods/HiddenFoods_analysis/krk2_genome_length_normalized_abundances_July2019/species&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Then we can import the abundance table files:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># --&gt; you can edit the files, eg remove last column &#39;Detail!</span>
<span class="c1"># dataframes in R must do not allow identical row names, force row names as numbering by using row.names=NULL.</span>
<span class="c1">#EOF issue with single quotes (&#39;) in species names, use read.delim</span>
HiddenFoods_Jan2019.krk2.norm.species <span class="o">=</span> read.delim<span class="p">(</span><span class="s">&quot;HiddenFoods_Jan2019.krk2.rcf.abundance.species.norm2.final&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">,</span> fill<span class="o">=</span><span class="bp">T</span><span class="p">,</span> row.names<span class="o">=</span><span class="kc">NULL</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
HiddenFoods_May2019.krk2.norm.species <span class="o">=</span> read.delim<span class="p">(</span><span class="s">&quot;HiddenFoods_May2019.krk2.rcf.abundance.species.norm2.final&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">,</span> fill<span class="o">=</span><span class="bp">T</span><span class="p">,</span> row.names<span class="o">=</span><span class="kc">NULL</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
Warinner2014.krk2.norm.species <span class="o">=</span> read.delim<span class="p">(</span><span class="s">&quot;Warinner2014.krk2.rcf.abundance.species.norm2.final&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">,</span> fill<span class="o">=</span><span class="bp">T</span><span class="p">,</span> row.names<span class="o">=</span><span class="kc">NULL</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
Velsko_modern_calculus.krk2.norm.species <span class="o">=</span> read.delim<span class="p">(</span><span class="s">&quot;Velsko_modern_calculus.krk2.rcf.abundance.species.norm2.final&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">,</span> fill<span class="o">=</span><span class="bp">T</span><span class="p">,</span> row.names<span class="o">=</span><span class="kc">NULL</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
Velsko_ancient_calculus_total.krk2.norm.species <span class="o">=</span> read.delim<span class="p">(</span><span class="s">&quot;Velsko_ancient_calculus_total.krk2.rcf.abundance.species.norm2.final&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">,</span> fill<span class="o">=</span><span class="bp">T</span><span class="p">,</span> row.names<span class="o">=</span><span class="kc">NULL</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
Plaque.krk2.norm.species <span class="o">=</span> read.delim<span class="p">(</span><span class="s">&quot;Plaque.krk2.rcf.abundance.species.norm2.final&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">,</span> fill<span class="o">=</span><span class="bp">T</span><span class="p">,</span> row.names<span class="o">=</span><span class="kc">NULL</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
Mann2018.krk2.norm.species <span class="o">=</span> read.delim<span class="p">(</span><span class="s">&quot;Mann2018.krk2.rcf.abundance.species.norm2.final&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">,</span> fill<span class="o">=</span><span class="bp">T</span><span class="p">,</span> row.names<span class="o">=</span><span class="kc">NULL</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
Baboons_Ozga.krk2.norm.species <span class="o">=</span> read.delim<span class="p">(</span><span class="s">&quot;Baboons_Ozga.krk2.rcf.abundance.species.norm2.final&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">,</span> fill<span class="o">=</span><span class="bp">T</span><span class="p">,</span> row.names<span class="o">=</span><span class="kc">NULL</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
Weyrich2017.krk2.norm.species <span class="o">=</span> read.delim<span class="p">(</span><span class="s">&quot;Weyrich2017.krk2.rcf.abundance.species.norm2.final&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">,</span> fill<span class="o">=</span><span class="bp">T</span><span class="p">,</span> row.names<span class="o">=</span><span class="kc">NULL</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
Baboons_Egypt_all.krk2.norm.species <span class="o">=</span> read.delim<span class="p">(</span><span class="s">&quot;Baboons_Egypt_all.krk2.rcf.abundance.species.flt10.teeth.env.norm2.final&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">,</span> fill<span class="o">=</span><span class="bp">T</span><span class="p">,</span> row.names<span class="o">=</span><span class="kc">NULL</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
Baboons_Egypt_nofilter.krk2.norm.species <span class="o">=</span> read.delim<span class="p">(</span><span class="s">&quot;Baboons_Egypt_nofilter.krk2.rcf.abundance.species.norm2.final&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">,</span> fill<span class="o">=</span><span class="bp">T</span><span class="p">,</span> row.names<span class="o">=</span><span class="kc">NULL</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
Baboons_Egypt_teeth.krk2.norm.species <span class="o">=</span> read.delim<span class="p">(</span><span class="s">&quot;Baboons_Egypt_teeth.krk2.rcf.abundance.species.norm2.final&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">,</span> fill<span class="o">=</span><span class="bp">T</span><span class="p">,</span> row.names<span class="o">=</span><span class="kc">NULL</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
Skin.krk2.norm.species <span class="o">=</span> read.delim<span class="p">(</span><span class="s">&quot;Skin.krk2.rcf.abundance.species.norm2.final&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">,</span> fill<span class="o">=</span><span class="bp">T</span><span class="p">,</span> row.names<span class="o">=</span><span class="kc">NULL</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
ObregonTito_gut.krk2.norm.species <span class="o">=</span> read.delim<span class="p">(</span><span class="s">&quot;ObregonTito_gut.krk2.rcf.abundance.species.norm2.final&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">,</span> fill<span class="o">=</span><span class="bp">T</span><span class="p">,</span> row.names<span class="o">=</span><span class="kc">NULL</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
Soil.krk2.norm.species <span class="o">=</span> read.delim<span class="p">(</span><span class="s">&quot;Soil.krk2.rcf.abundance.species.norm2.final&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">,</span> fill<span class="o">=</span><span class="bp">T</span><span class="p">,</span> row.names<span class="o">=</span><span class="kc">NULL</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
Chimps.krk2.norm.species <span class="o">=</span> read.delim<span class="p">(</span><span class="s">&quot;Chimps_Ozga.krk2.rcf.abundance.species.norm2.final&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">,</span> fill<span class="o">=</span><span class="bp">T</span><span class="p">,</span> row.names<span class="o">=</span><span class="kc">NULL</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
Brealey_animals.krk2.norm.species <span class="o">=</span> read.delim<span class="p">(</span><span class="s">&quot;Brealey_animals.krk2.rcf.abundance.species.norm2.final&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">,</span> fill<span class="o">=</span><span class="bp">T</span><span class="p">,</span> row.names<span class="o">=</span><span class="kc">NULL</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
Brealey_animals.krk2.norm.species <span class="o">=</span> Brealey_animals.krk2.norm.species<span class="p">[,</span><span class="o">-</span><span class="kt">c</span><span class="p">(</span><span class="m">2</span><span class="o">:</span><span class="m">5</span><span class="p">,</span><span class="m">7</span><span class="p">,</span><span class="m">8</span><span class="p">)]</span>           <span class="c1">#remove Blanks from filtrated Brealey, Gb1reg/flt, Gb2reg/flt</span>
Eisenhofer_Japan_flt.krk2.norm.species <span class="o">=</span> read.table<span class="p">(</span><span class="s">&quot;Eisenhofer_Japan.krk2.rcf.abundance.species.norm2.final.flt&quot;</span><span class="p">,</span> header<span class="o">=</span><span class="bp">T</span><span class="p">,</span> fill<span class="o">=</span><span class="bp">T</span><span class="p">,</span> row.names<span class="o">=</span><span class="kc">NULL</span><span class="p">,</span> sep<span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
Eisenhofer_Japan_flt.krk2.norm.species <span class="o">=</span> Eisenhofer_Japan_flt.krk2.norm.species<span class="p">[,</span><span class="o">-</span><span class="kt">c</span><span class="p">(</span><span class="m">19</span><span class="p">,</span><span class="m">20</span><span class="p">)]</span>   <span class="c1">#remove EBC-2 in  Eisenhofer which has no classified species, and Details column.</span>
</pre></div>
</div>
<div class="section" id="barplot">
<h2>8.1. Barplot<a class="headerlink" href="#barplot" title="Permalink to this headline">Â¶</a></h2>
</div>
<div class="section" id="upgma">
<h2>8.2. UPGMA<a class="headerlink" href="#upgma" title="Permalink to this headline">Â¶</a></h2>
</div>
<div class="section" id="nmds">
<h2>8.3. nMDS<a class="headerlink" href="#nmds" title="Permalink to this headline">Â¶</a></h2>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="11_old_material.html" class="btn btn-neutral float-right" title="9. Material from previous courses" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="9_DIY.html" class="btn btn-neutral float-left" title="7. DO-IT-YOURSELF" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2020, Amine Namouchi &amp; Claudio Ottoni

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
    <!-- Theme Analytics -->
    <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-XXXXXXX-1', 'auto');
    ga('send', 'pageview');
    </script>

    
   

</body>
</html>