

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>2. First step to set-up your machine &mdash; Paleogenomics at Tor Vergata 0.1.0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="3. Useful commands" href="2_Useful_commands.html" />
    <link rel="prev" title="1. List of Tools" href="1_ListTools.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> Paleogenomics at Tor Vergata
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="1_ListTools.html">1. List of Tools</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">2. First step to set-up your machine</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#set-up-your-terminal-machine">2.1. Set-up your terminal machine</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#set-up-for-windows-users">2.1.1. Set-up for Windows users</a></li>
<li class="toctree-l3"><a class="reference internal" href="#set-up-for-mac-osx-users">2.1.2. Set-up for Mac OSX users</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#install-miniconda-to-work-on-your-local-machine">2.2. Install Miniconda to work on your local machine</a></li>
<li class="toctree-l2"><a class="reference internal" href="#create-an-environment-with-conda">2.3. Create an environment with Conda</a></li>
<li class="toctree-l2"><a class="reference internal" href="#install-packages-inside-the-environment">2.4. Install packages inside the environment</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="2_Useful_commands.html">3. Useful commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="3_Slurm_HPC.html">4. Working on HPC with Slurm</a></li>
<li class="toctree-l1"><a class="reference internal" href="3_ReadsFiltering.html">5. Quality filtering of reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="4_Metagenomics_v2.html">6. Metagenomic screening of shotgun data</a></li>
<li class="toctree-l1"><a class="reference internal" href="5_ReadsMapping_v2.html">7. Alignment of reads to a reference genome</a></li>
<li class="toctree-l1"><a class="reference internal" href="6_VariantsCall.html">8. Variant calling and visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="7_Filtering_SNPs.html">9. Filtering, annotating and combining SNPs</a></li>
<li class="toctree-l1"><a class="reference internal" href="8_HandsOn.html">10. Hands-on</a></li>
<li class="toctree-l1"><a class="reference internal" href="9_old_material.html">11. Material from previous courses</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Paleogenomics at Tor Vergata</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li><span class="section-number">2. </span>First step to set-up your machine</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/2_Terminal_setup.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="first-step-to-set-up-your-machine">
<h1><span class="section-number">2. </span>First step to set-up your machine<a class="headerlink" href="#first-step-to-set-up-your-machine" title="Permalink to this headline">¶</a></h1>
<div class="section" id="set-up-your-terminal-machine">
<h2><span class="section-number">2.1. </span>Set-up your terminal machine<a class="headerlink" href="#set-up-your-terminal-machine" title="Permalink to this headline">¶</a></h2>
<div class="section" id="set-up-for-windows-users">
<h3><span class="section-number">2.1.1. </span>Set-up for Windows users<a class="headerlink" href="#set-up-for-windows-users" title="Permalink to this headline">¶</a></h3>
<p>If you a have a Windows-based computer you must download the app of <a class="reference external" href="https://www.microsoft.com/it-it/p/ubuntu/9nblggh4msv6?activetab=pivot:overviewtab">Ubuntu</a> in the Microsoft Store.</p>
<p>After that you have to enable the use of Linux in Windows:
- From the Start menu, search “Programmi e funzionalità”.
- Go to the menu “Attivazione/disattivazione delle funzionalità Windows”.
- Check the box “Sottosistema Windows per Linux” to enable it (see the figure below).</p>
<img alt="_images/Sottosistema-Linux.jpg" src="_images/Sottosistema-Linux.jpg" />
<p>Now you can open the app Ubuntu and use the textual shell <strong>bash</strong> in the terminal.
<a class="reference external" href="https://drive.google.com/file/d/18v7EoQcg1ff3My4bypvjzkMUiIFrMkT2/view?usp=sharing">Here</a> you can find a quick reference to the basic commands in bash.</p>
</div>
<div class="section" id="set-up-for-mac-osx-users">
<h3><span class="section-number">2.1.2. </span>Set-up for Mac OSX users<a class="headerlink" href="#set-up-for-mac-osx-users" title="Permalink to this headline">¶</a></h3>
<p>Mac OS has already a Terminal app. Open it and make sure that you have the Command Line Tools installed by typing the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">xcode</span><span class="o">-</span><span class="n">select</span> <span class="o">--</span><span class="n">install</span>
</pre></div>
</div>
<p>The Command Line Tools package will be installed in your Mac, so that you can use the standard commands to work in the shell.</p>
</div>
</div>
<div class="section" id="install-miniconda-to-work-on-your-local-machine">
<h2><span class="section-number">2.2. </span>Install Miniconda to work on your local machine<a class="headerlink" href="#install-miniconda-to-work-on-your-local-machine" title="Permalink to this headline">¶</a></h2>
<p><a class="reference external" href="https://drive.google.com/file/d/18v7EoQcg1ff3My4bypvjzkMUiIFrMkT2/view?usp=sharing">Here</a> you can find a quick reference to the basic commands in bash.</p>
<p>To work locally (on your computer) you will have to install several programs and software packages. This can be done with a package manager <strong>Miniconda</strong>.
For Windows users, to install Miniconda type the following commands:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">wget</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">repo</span><span class="o">.</span><span class="n">anaconda</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">miniconda</span><span class="o">/</span><span class="n">Miniconda3</span><span class="o">-</span><span class="n">py39_4</span><span class="mf">.10.3</span><span class="o">-</span><span class="n">Linux</span><span class="o">-</span><span class="n">x86_64</span><span class="o">.</span><span class="n">sh</span>
<span class="n">bash</span> <span class="n">Miniconda3</span><span class="o">-</span><span class="n">py39_4</span><span class="mf">.10.3</span><span class="o">-</span><span class="n">Linux</span><span class="o">-</span><span class="n">x86_64</span><span class="o">.</span><span class="n">sh</span>
</pre></div>
</div>
<p>for Mac-OS users type:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">curl</span> <span class="o">-</span><span class="n">O</span> <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">repo</span><span class="o">.</span><span class="n">anaconda</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">miniconda</span><span class="o">/</span><span class="n">Miniconda3</span><span class="o">-</span><span class="n">latest</span><span class="o">-</span><span class="n">MacOSX</span><span class="o">-</span><span class="n">x86_64</span><span class="o">.</span><span class="n">sh</span>
<span class="n">bash</span> <span class="n">Miniconda3</span><span class="o">-</span><span class="n">latest</span><span class="o">-</span><span class="n">MacOSX</span><span class="o">-</span><span class="n">x86_64</span><span class="o">.</span><span class="n">sh</span>
</pre></div>
</div>
<p>Follow the procedure (type always <strong>yes</strong> if asked) and when asked about executing <strong>conda init</strong> type <strong>yes</strong>.
Miniconda automatically installs also <strong>Python 3</strong>.</p>
</div>
<div class="section" id="create-an-environment-with-conda">
<h2><span class="section-number">2.3. </span>Create an environment with Conda<a class="headerlink" href="#create-an-environment-with-conda" title="Permalink to this headline">¶</a></h2>
<p>A conda environment is a directory that contains a specific collection of conda packages that you have installed. For example, you may have one environment with Python 2 and its dependencies, and another environment with Python 3.
Environments and all programs inside an environment are independent from programs belonging to another environment.
Find here more information on Conda <a class="reference external" href="https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#">environments</a>.</p>
<p>Here we will create an environment called <strong>bioinf</strong>, where we will install most of the programs that we will use.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">create</span> <span class="o">--</span><span class="n">name</span> <span class="n">bioinf</span>
</pre></div>
</div>
<p>After creating the environment, activate it by typing the following command (and this must be don everytime you want to activate the environment)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">activate</span> <span class="n">bioinf</span>
</pre></div>
</div>
<p>To deactivate the environment</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">deactivate</span>
</pre></div>
</div>
</div>
<div class="section" id="install-packages-inside-the-environment">
<h2><span class="section-number">2.4. </span>Install packages inside the environment<a class="headerlink" href="#install-packages-inside-the-environment" title="Permalink to this headline">¶</a></h2>
<p>You will install most of the programs that we will use in the <strong>bioinf</strong> environment, so first of all activate the environment.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">activate</span> <span class="n">bioinf</span>
</pre></div>
</div>
<p>Install FastQC:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">bioconda</span> <span class="n">fastqc</span>
</pre></div>
</div>
<p>Install Adapter Removal</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">bioconda</span> <span class="n">adapterremoval</span>
</pre></div>
</div>
<p>Install BWA:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">bioconda</span> <span class="n">bwa</span>
</pre></div>
</div>
<p>Install Picard:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">bioconda</span> <span class="n">picard</span>
</pre></div>
</div>
<p>Install Samtools:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">bioconda</span> <span class="n">samtools</span>
</pre></div>
</div>
<p>If you have troubles with the installation of Samtools try these commands:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">config</span> <span class="o">--</span><span class="n">add</span> <span class="n">channels</span> <span class="n">bioconda</span>
<span class="n">conda</span> <span class="n">config</span> <span class="o">--</span><span class="n">add</span> <span class="n">channels</span> <span class="n">conda</span><span class="o">-</span><span class="n">forge</span>
<span class="n">conda</span> <span class="n">install</span> <span class="n">samtools</span><span class="o">==</span><span class="mf">1.11</span>
</pre></div>
</div>
<p>Install Dedup:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">conda</span> <span class="n">install</span> <span class="o">-</span><span class="n">c</span> <span class="n">bioconda</span> <span class="n">dedup</span>
</pre></div>
</div>
<p>We will provide a version of <strong>GATK 3.8</strong> (file <code class="docutils literal notranslate"><span class="pre">GenomeAnalysisTK.jar</span></code>).
Download the <a class="reference external" href="https://software.broadinstitute.org/software/igv/download">IGV Genome browser</a> on you computer, just press the link and download the version for your computed (Mac-OS or Windows).</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="2_Useful_commands.html" class="btn btn-neutral float-right" title="3. Useful commands" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="1_ListTools.html" class="btn btn-neutral float-left" title="1. List of Tools" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2021, Claudio Ottoni

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
    <!-- Theme Analytics -->
    <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-XXXXXXX-1', 'auto');
    ga('send', 'pageview');
    </script>

    
   

</body>
</html>