

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>3. Metagenomic screening of shotgun data &mdash; Physalia Paleogenomics 0.1.0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="4. Alignment of reads to a reference genome" href="4_ReadsMapping_v2.html" />
    <link rel="prev" title="2. Quality filtering of reads" href="2_ReadsFiltering.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> Physalia Paleogenomics
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="1_ListTools.html">1. List of Tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="2_ReadsFiltering.html">2. Quality filtering of reads</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">3. Metagenomic screening of shotgun data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#kraken-2">3.1. Kraken 2</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#minikraken2">3.1.1. Minikraken2</a></li>
<li class="toctree-l3"><a class="reference internal" href="#visualization-of-data-with-krona">3.1.2. Visualization of data with Krona</a></li>
<li class="toctree-l3"><a class="reference internal" href="#kraken-2-custom-database">3.1.3. Kraken 2 Custom Database</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#bracken">3.2. Bracken</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#bracken-from-minikraken-2">3.2.1. Bracken from Minikraken 2</a></li>
<li class="toctree-l3"><a class="reference internal" href="#bracken-from-custom-databases">3.2.2. Bracken from Custom databases</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#mataphlan-2">3.3. Mataphlan 2</a></li>
<li class="toctree-l2"><a class="reference internal" href="#other-tools">3.4. Other tools</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="4_ReadsMapping_v2.html">4. Alignment of reads to a reference genome</a></li>
<li class="toctree-l1"><a class="reference internal" href="7_VariantsCall.html">5. Variant calling and visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="8_Filtering_SNPs.html">6. Filtering, annotating and combining SNPs</a></li>
<li class="toctree-l1"><a class="reference internal" href="9_DIY.html">7. DO-IT-YOURSELF</a></li>
<li class="toctree-l1"><a class="reference internal" href="10_R_session.html">8. R session</a></li>
<li class="toctree-l1"><a class="reference internal" href="11_old_material.html">9. Material from previous courses</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Physalia Paleogenomics</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li><span class="section-number">3. </span>Metagenomic screening of shotgun data</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/3_Metagenomics_v2.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="metagenomic-screening-of-shotgun-data">
<h1><span class="section-number">3. </span>Metagenomic screening of shotgun data<a class="headerlink" href="#metagenomic-screening-of-shotgun-data" title="Permalink to this headline">¶</a></h1>
<div class="section" id="kraken-2">
<h2><span class="section-number">3.1. </span>Kraken 2<a class="headerlink" href="#kraken-2" title="Permalink to this headline">¶</a></h2>
<p>In this hands-on session we will use Kraken 2 to screen the metagenomic content of an ancient sample. The first version of Kraken (see <a class="reference internal" href="11_old_material.html#kraken-label"><span class="std std-ref">Kraken</span></a>) uses a large indexed and sorted list of k-mer/LCA pairs as its database, which may be problematic for users due to the large memory requirements. For this reason <a class="reference external" href="https://ccb.jhu.edu/software/kraken2/index.shtml">Kraken 2</a> was developed.
Kraken 2 is fast and requires less memory, BUT false positive errors occur in less than 1% of queries (confidence scoring thresholds can be used to call out to detect them).
The first version of Kraken use a large indexed and sorted list of k-mer/LCA pairs as its database, which may be problematic for users due to the large memory requirements. For this reason <a class="reference external" href="https://ccb.jhu.edu/software/kraken2/index.shtml">Kraken 2</a> was developed.
Kraken 2 is fast and requires less memory, BUT the database false positive errors occur in less than 1% of queries (confidence scoring thresholds can be used to call out to detect them).
The default (or standard) database size is 29 GB (as of Jan. 2018, in Kraken 1 the standard database is about 200 GB!), and you will need slightly more than that in RAM if you want to build it.
By default, Kraken 2 will attempt to use the dustmasker or  segmasker programs provided as part of NCBI’s BLAST suite to mask low-complexity regions.</p>
<p>A Kraken 2 database is a directory containing at least 3 files:</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>hash.k2d</strong>: Contains the minimizer to taxon mappings</p></li>
<li><p><strong>opts.k2d</strong>: Contains information about the options used to build the database</p></li>
<li><p><strong>taxo.k2d</strong>: Contains taxonomy information used to build the database</p></li>
</ul>
</div></blockquote>
<p>Other files may also be present as part of the database build process, and can, if desired, be removed after a successful build of the database.</p>
<blockquote>
<div></div></blockquote>
<div class="section" id="minikraken2">
<h3><span class="section-number">3.1.1. </span>Minikraken2<a class="headerlink" href="#minikraken2" title="Permalink to this headline">¶</a></h3>
<p>We will first use a pre-built 8 GB Kraken database, called <a class="reference external" href="https://ccb.jhu.edu/software/kraken2/index.shtml?t=downloads">Minikraken2</a>, constructed from complete dusted bacterial, archaeal, and viral genomes in RefSeq. You can download the pre-built <a class="reference external" href="https://ccb.jhu.edu/software/kraken2/index.shtml?t=downloads">Minikraken2</a> database from the program’s website with <code class="docutils literal notranslate"><span class="pre">wget</span></code>, and extract the archive content with <code class="docutils literal notranslate"><span class="pre">tar</span></code>:
You can download the pre-built <a class="reference external" href="https://ccb.jhu.edu/software/kraken2/index.shtml?t=downloads">Minikraken2</a> database (containing bacteria, viral and archaea sequences) from the website with <code class="docutils literal notranslate"><span class="pre">wget</span></code>, and extract the archive content with <code class="docutils literal notranslate"><span class="pre">tar</span></code>:</p>
<blockquote>
<div></div></blockquote>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">wget</span> <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">ccb</span><span class="o">.</span><span class="n">jhu</span><span class="o">.</span><span class="n">edu</span><span class="o">/</span><span class="n">pub</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">kraken2_dbs</span><span class="o">/</span><span class="n">minikraken2_v1_8GB_201904_UPDATE</span><span class="o">.</span><span class="n">tgz</span>
<span class="n">tar</span> <span class="o">-</span><span class="n">xvzf</span> <span class="n">minikraken2_v1_8GB_201904_UPDATE</span><span class="o">.</span><span class="n">tgz</span>
</pre></div>
</div>
<p>To classify the reads in a <code class="docutils literal notranslate"><span class="pre">fastq</span></code> file against the Minikraken2 database, you can run this command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">kraken2</span> <span class="o">--</span><span class="n">db</span> <span class="n">minikraken2_v1_8GB</span> <span class="n">filename</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="o">--</span><span class="n">gzip</span><span class="o">-</span><span class="n">compressed</span> <span class="o">--</span><span class="n">output</span> <span class="n">filename</span><span class="o">.</span><span class="n">kraken</span> <span class="o">--</span><span class="n">report</span> <span class="n">filename</span><span class="o">.</span><span class="n">kraken</span><span class="o">.</span><span class="n">report</span>
</pre></div>
</div>
<p>Some of the options available in Kraken 2:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 29%" />
<col style="width: 71%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Option</p></th>
<th class="head"><p>Function</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>–use-names</strong></p></td>
<td><p>Print scientific names instead of just taxids</p></td>
</tr>
<tr class="row-odd"><td><p><strong>–gzip-compressed</strong></p></td>
<td><p>Input is gzip compressed</p></td>
</tr>
<tr class="row-even"><td><p><strong>–report</strong> &lt;string&gt;</p></td>
<td><p>Print a report with aggregrate counts/clade to file</p></td>
</tr>
<tr class="row-odd"><td><p><strong>–threads</strong></p></td>
<td><p>Number of threads (default: 1)</p></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul class="simple">
<li><p>In Kraken 2 you can generate the reports file by typing the <code class="docutils literal notranslate"><span class="pre">--report</span></code> option (followed by a name for the report file to generate) in the command used for the classification.</p></li>
<li><p>In order to run later Krona, the Kraken output file must contain taxids, and not scientific names. So if you want to run Krona do not call the option <code class="docutils literal notranslate"><span class="pre">--use-names</span></code>.</p></li>
</ul>
</div>
<p>The Kraken 2 report format, like Kraken 1, is tab-delimited with one line per taxon. There are six fields, from left to right:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>Percentage of fragments covered by the clade rooted at this taxon</p></li>
<li><p>Number of fragments covered by the clade rooted at this taxon</p></li>
<li><p>Number of fragments assigned directly to this taxon</p></li>
<li><p>A rank code, indicating (U)nclassified, (R)oot, (D)omain, (K)ingdom,
(P)hylum, (C)lass, (O)rder, (F)amily, (G)enus, or (S)pecies.
Taxa that are not at any of these 10 ranks have a rank code that is
formed by using the rank code of the closest ancestor rank with
a number indicating the distance from that rank.  E.g., “G2” is a
rank code indicating a taxon is between genus and species and the
grandparent taxon is at the genus rank.</p></li>
<li><p>NCBI taxonomic ID number</p></li>
<li><p>Indented scientific name</p></li>
</ol>
</div></blockquote>
</div>
<div class="section" id="visualization-of-data-with-krona">
<span id="krona-label"></span><h3><span class="section-number">3.1.2. </span>Visualization of data with Krona<a class="headerlink" href="#visualization-of-data-with-krona" title="Permalink to this headline">¶</a></h3>
<p>Finally, we can visualize the results of the Kraken2 analysis with <a class="reference external" href="https://github.com/marbl/Krona/wiki">Krona</a>, which disaplys hierarchical data (like taxonomic assignation) in multi-layerd pie charts.
The interactive charts created by Krona are in the <code class="docutils literal notranslate"><span class="pre">html</span></code> format and can be viewed with any web browser.
We will convert the kraken output in html format using the program <code class="docutils literal notranslate"><span class="pre">ktImportTaxonomy</span></code>, which parses the information relative to the query ID and the taxonomy ID.</p>
<blockquote>
<div></div></blockquote>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ktImportTaxonomy</span> <span class="o">-</span><span class="n">q</span> <span class="mi">2</span> <span class="o">-</span><span class="n">t</span> <span class="mi">3</span> <span class="n">filename</span><span class="o">.</span><span class="n">kraken</span> <span class="o">-</span><span class="n">o</span> <span class="n">filename</span><span class="o">.</span><span class="n">kraken</span><span class="o">.</span><span class="n">html</span>
</pre></div>
</div>
<p>Some of the options available in ktImportTaxonomy:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 27%" />
<col style="width: 73%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Option</p></th>
<th class="head"><p>Function</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>-q</strong> &lt;integer&gt;</p></td>
<td><p>Column of input files to use as query ID.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>-t</strong> &lt;integer&gt;</p></td>
<td><p>Column of input files to use as taxonomy ID.</p></td>
</tr>
<tr class="row-even"><td><p><strong>-o</strong> &lt;string&gt;</p></td>
<td><p>Output file name.</p></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If you want to analyze multiple kraken files from various samples you view the results in one single html file running <code class="docutils literal notranslate"><span class="pre">ktImportTaxonomy</span></code> as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ktImportTaxonomy</span> <span class="o">-</span><span class="n">q</span> <span class="mi">2</span> <span class="o">-</span><span class="n">t</span> <span class="mi">3</span> <span class="n">filename_1</span><span class="o">.</span><span class="n">kraken</span> <span class="n">filename_2</span><span class="o">.</span><span class="n">kraken</span> <span class="o">...</span> <span class="n">filename_n</span><span class="o">.</span><span class="n">kraken</span> <span class="o">-</span><span class="n">o</span> <span class="n">all_samples</span><span class="o">.</span><span class="n">kraken</span><span class="o">.</span><span class="n">html</span>
</pre></div>
</div>
</div>
<img alt="_images/krona.png" src="_images/krona.png" />
</div>
<div class="section" id="kraken-2-custom-database">
<h3><span class="section-number">3.1.3. </span>Kraken 2 Custom Database<a class="headerlink" href="#kraken-2-custom-database" title="Permalink to this headline">¶</a></h3>
<p>We have already created a custom database to use in this hands-on session so we can go straight to the classification (step 4). However, we report here all the commands to build a Kraken2 database (steps 1-3).</p>
<ol class="arabic">
<li><p>The first step is to create a new folder that will contain your custom database (choose an appropriate name for the folder-database, here we will call it <code class="docutils literal notranslate"><span class="pre">CustomDB</span></code>).
Then we have to install a taxonomy. Usually, you will use the NCBI taxonomy. The following command will download in the folder <code class="docutils literal notranslate"><span class="pre">/taxonomy</span></code> the accession number to taxon maps, as well as the taxonomic name and tree information from NCBI:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">kraken2</span><span class="o">-</span><span class="n">build</span> <span class="o">--</span><span class="n">download</span><span class="o">-</span><span class="n">taxonomy</span> <span class="o">--</span><span class="n">db</span> <span class="n">CustomDB</span>
</pre></div>
</div>
</li>
<li><p>Install one or more reference libraries. Several sets of standard genomes (or proteins) are available, which are constantly updated (see also the Kraken <a class="reference external" href="https://ccb.jhu.edu/software/kraken2/index.shtml?t=manual#installation)">website</a>).</p>
<blockquote>
<div><ul class="simple">
<li><p>archaea: RefSeq complete archaeal genomes/proteins</p></li>
<li><p>bacteria: RefSeq complete bacterial genomes/proteins</p></li>
<li><p>plasmid: RefSeq plasmid nucleotide/protein sequences</p></li>
<li><p>viral: RefSeq complete viral genomes/proteins</p></li>
<li><p>human: GRCh38 human genome/proteins</p></li>
<li><p>fungi: RefSeq complete fungal genomes/proteins</p></li>
<li><p>plant: RefSeq complete plant genomes/proteins</p></li>
<li><p>protozoa: RefSeq complete protozoan genomes/proteins</p></li>
<li><p>nr: NCBI non-redundant protein database</p></li>
<li><p>nt: NCBI non-redundant nucleotide database</p></li>
<li><p>env_nr: NCBI non-redundant protein database with sequences from large environmental sequencing projects</p></li>
<li><p>env_nt: NCBI non-redundant nucleotide database with sequences from large environmental sequencing projects</p></li>
<li><p>UniVec: NCBI-supplied database of vector, adapter, linker, and primer sequences that may be contaminating sequencing projects and/or assemblies</p></li>
<li><p>UniVec_Core: A subset of UniVec chosen to minimize false positive hits to the vector database</p></li>
</ul>
</div></blockquote>
<p>You can select as many libraries as you want and run the following command, which will download the reference sequences in the folder <code class="docutils literal notranslate"><span class="pre">/library</span></code>, as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">kraken2</span><span class="o">-</span><span class="n">build</span> <span class="o">--</span><span class="n">download</span><span class="o">-</span><span class="n">library</span> <span class="n">bacteria</span> <span class="o">--</span><span class="n">db</span> <span class="n">CustomDB</span>
<span class="n">kraken2</span><span class="o">-</span><span class="n">build</span> <span class="o">--</span><span class="n">download</span><span class="o">-</span><span class="n">library</span> <span class="n">viral</span> <span class="o">--</span><span class="n">db</span> <span class="n">CustomDB</span>
<span class="n">kraken2</span><span class="o">-</span><span class="n">build</span> <span class="o">--</span><span class="n">download</span><span class="o">-</span><span class="n">library</span> <span class="n">plasmid</span> <span class="o">--</span><span class="n">db</span> <span class="n">CustomDB</span>
<span class="n">kraken2</span><span class="o">-</span><span class="n">build</span> <span class="o">--</span><span class="n">download</span><span class="o">-</span><span class="n">library</span> <span class="n">fungi</span> <span class="o">--</span><span class="n">db</span> <span class="n">CustomDB</span>
</pre></div>
</div>
<p>In a custom database, you can add as many <code class="docutils literal notranslate"><span class="pre">fasta</span></code> sequences as you like. For exmaple, you can download the organelle genomes in <code class="docutils literal notranslate"><span class="pre">fasta</span></code> files from the <a class="reference external" href="https://www.ncbi.nlm.nih.gov/genome/organelle/">RefSeq</a> website with the commands <code class="docutils literal notranslate"><span class="pre">wget</span></code>:</p>
<blockquote>
<div></div></blockquote>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">wget</span> <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">ncbi</span><span class="o">.</span><span class="n">nlm</span><span class="o">.</span><span class="n">nih</span><span class="o">.</span><span class="n">gov</span><span class="o">/</span><span class="n">refseq</span><span class="o">/</span><span class="n">release</span><span class="o">/</span><span class="n">mitochondrion</span><span class="o">/</span><span class="n">mitochondrion</span><span class="o">.</span><span class="mf">1.1</span><span class="o">.</span><span class="n">genomic</span><span class="o">.</span><span class="n">fna</span><span class="o">.</span><span class="n">gz</span>
<span class="n">wget</span> <span class="n">ftp</span><span class="p">:</span><span class="o">//</span><span class="n">ftp</span><span class="o">.</span><span class="n">ncbi</span><span class="o">.</span><span class="n">nlm</span><span class="o">.</span><span class="n">nih</span><span class="o">.</span><span class="n">gov</span><span class="o">/</span><span class="n">refseq</span><span class="o">/</span><span class="n">release</span><span class="o">/</span><span class="n">mitochondrion</span><span class="o">/</span><span class="n">mitochondrion</span><span class="o">.</span><span class="mf">2.1</span><span class="o">.</span><span class="n">genomic</span><span class="o">.</span><span class="n">fna</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>The downloaded files are in compressed in the <code class="docutils literal notranslate"><span class="pre">gz</span></code> format. To unzip them run the <code class="docutils literal notranslate"><span class="pre">gunzip</span></code> command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">gunzip</span> <span class="n">mitochondrion</span><span class="o">.</span><span class="mf">1.1</span><span class="o">.</span><span class="n">genomic</span><span class="o">.</span><span class="n">fna</span><span class="o">.</span><span class="n">gz</span>
<span class="n">gunzip</span> <span class="n">mitochondrion</span><span class="o">.</span><span class="mf">2.1</span><span class="o">.</span><span class="n">genomic</span><span class="o">.</span><span class="n">fna</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>Then you can add the <code class="docutils literal notranslate"><span class="pre">fasta</span></code> files to your library, as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">kraken2</span><span class="o">-</span><span class="n">build</span> <span class="o">--</span><span class="n">add</span><span class="o">-</span><span class="n">to</span><span class="o">-</span><span class="n">library</span> <span class="n">mitochondrion</span><span class="o">.</span><span class="mf">1.1</span><span class="o">.</span><span class="n">genomic</span><span class="o">.</span><span class="n">fna</span> <span class="o">--</span><span class="n">db</span> <span class="n">CustomDB</span>
<span class="n">kraken2</span><span class="o">-</span><span class="n">build</span> <span class="o">--</span><span class="n">add</span><span class="o">-</span><span class="n">to</span><span class="o">-</span><span class="n">library</span> <span class="n">mitochondrion</span><span class="o">.</span><span class="mf">2.1</span><span class="o">.</span><span class="n">genomic</span><span class="o">.</span><span class="n">fna</span> <span class="o">--</span><span class="n">db</span> <span class="n">CustomDB</span>
</pre></div>
</div>
</li>
<li><p>Once your library is finalized, you need to build the database (here we set a maximum database size of 8 GB (you must indicate it in bytes!) with the <code class="docutils literal notranslate"><span class="pre">--max-db-size</span></code> option).</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">kraken2</span><span class="o">-</span><span class="n">build</span> <span class="o">--</span><span class="n">build</span> <span class="o">--</span><span class="nb">max</span><span class="o">-</span><span class="n">db</span><span class="o">-</span><span class="n">size</span> <span class="mi">8000000000</span> <span class="o">--</span><span class="n">db</span> <span class="n">CustomDB</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>Kraken 2 uses two programs to perform low-complexity sequence masking, both available from NCBI: <code class="docutils literal notranslate"><span class="pre">dustmasker</span></code>, for nucleotide sequences, and  <code class="docutils literal notranslate"><span class="pre">segmasker</span></code>, for amino acid sequences.
These programs are available as part of the NCBI BLAST+ suite. If these programs are not installed on the local system and in the user’s PATH when trying to use <code class="docutils literal notranslate"><span class="pre">kraken2-build</span></code>, the database build will fail.
Users who do not wish to install these programs can use the <code class="docutils literal notranslate"><span class="pre">--no-masking</span></code> option to kraken2-build in conjunction with any of the <code class="docutils literal notranslate"><span class="pre">--download-library</span></code>, <code class="docutils literal notranslate"><span class="pre">--add-to-library</span></code>, or <code class="docutils literal notranslate"><span class="pre">--standard</span></code> options; use of the <code class="docutils literal notranslate"><span class="pre">--no-masking</span></code> option will skip masking of low-complexity sequences during the build of the Kraken 2 database.</p>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">kraken2-inspect</span></code> script allows users to gain information about the content of a Kraken 2 database. You can pipe the command to <code class="docutils literal notranslate"><span class="pre">head</span></code>, or <code class="docutils literal notranslate"><span class="pre">less</span></code>.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">kraken2</span><span class="o">-</span><span class="n">inspect</span> <span class="o">--</span><span class="n">db</span> <span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">dbfolder</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="mi">5</span>
</pre></div>
</div>
</li>
<li><p>Finally, we can run the classification of the reads against the custom database with the <code class="docutils literal notranslate"><span class="pre">kraken2</span></code> command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">kraken2</span> <span class="o">--</span><span class="n">db</span> <span class="n">CustomDB</span> <span class="n">filename</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="o">--</span><span class="n">gzip</span><span class="o">-</span><span class="n">compressed</span> <span class="o">--</span><span class="n">output</span> <span class="n">filename</span><span class="o">.</span><span class="n">kraken</span> <span class="o">--</span><span class="n">report</span> <span class="n">filename</span><span class="o">.</span><span class="n">report</span>
</pre></div>
</div>
</li>
<li><p>To visualize the results of the classification in multi-layerd pie charts, use <code class="docutils literal notranslate"><span class="pre">Krona</span></code>, as described in the section 3.1.2: <a class="reference internal" href="#visualization-of-data-with-krona">Visualization of data with Krona</a></p></li>
</ol>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Recently, a novel metagenomics classifier, <a class="reference external" href="https://github.com/fbreitwieser/krakenuniq">KrakenUniq</a>, has been developed to reduce false-positive identifications in metagenomic classification.
KrakenUniq combines the fast k-mer-based classification of Kraken with an efficient algorithm for assessing the coverage of unique k-mers found in each species in a dataset.
On various test datasets, KrakenUniq gives better recall and precision than other methods and effectively classifies and distinguishes pathogens with low abundance from false positives in infectious disease samples.</p>
<blockquote>
<div></div></blockquote>
</div>
</div>
</div>
<div class="section" id="bracken">
<h2><span class="section-number">3.2. </span>Bracken<a class="headerlink" href="#bracken" title="Permalink to this headline">¶</a></h2>
<p>Kraken and Kraken 2 classify reads to the best matching location in the taxonomic tree, but they do not estimate abundances of species.
To do that we will use Bracken (Bayesian Reestimation of Abundance with KrakEN), which estimates the number of reads originating from each species present in a sample. Bracken computes probabilities that describe how much sequence from each genome in the Kraken database is identical to other genomes in the database, and combine this information with the assignments for a particular sample to estimate abundance at the species level, the genus level, or above.
Bracken is compatible with both Kraken 1 and Kraken 2 (just note that the default kmer length is different, 31 in Kraken, 35 in Kraken 2).</p>
<div class="section" id="bracken-from-minikraken-2">
<h3><span class="section-number">3.2.1. </span>Bracken from Minikraken 2<a class="headerlink" href="#bracken-from-minikraken-2" title="Permalink to this headline">¶</a></h3>
<p>Prior to abundance estimation with Bracken, each reference genome contained in the Kraken database is divided into “read-length kmers”, and each of these read-length kmers are classified. To do that, we need the <strong>library</strong> and <strong>taxonomy</strong> of the kraken database that we used for classification. This will generate a kmer distribution file, for examples for read-length kmers of 100 nucleotides: <code class="docutils literal notranslate"><span class="pre">database100mers.kmer_distrib</span></code>.
The kmer distribution file will be used for the following bracken analysis.
As pre-built database, Minikraken already contains three kmer distribution files built at 100, 150, 200 kmers.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 5%" />
<col style="width: 95%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Option</p></th>
<th class="head"><p>Function</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>-i</strong> &lt;string&gt;</p></td>
<td><p>Kraken report input file.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>-o</strong> &lt;string&gt;</p></td>
<td><p>Output file name.</p></td>
</tr>
<tr class="row-even"><td><p><strong>-t</strong> &lt;string&gt;</p></td>
<td><p>Classification level [Default = ‘S’, Options = ‘D’,’P’,’C’,’O’,’F’,’G’,’S’]: it specifies the taxonomic rank to analyze. Each classification at this specified rank will receive an estimated number of reads belonging to that rank after abundance estimation.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>-l</strong> &lt;string&gt;</p></td>
<td><p>Threshold: it specifies the minimum number of reads required for a classification at the specified rank. Any classifications with less than the specified threshold will not receive additional reads from higher taxonomy levels when distributing reads for abundance estimation.</p></td>
</tr>
<tr class="row-even"><td><p><strong>-r</strong> &lt;string&gt;</p></td>
<td><p>Read-length kmer used for the generations of Bracken distribution files (in Minikraken 100, 150, 200)</p></td>
</tr>
</tbody>
</table>
<p>To run Bracken on one sample we use the following command, with a threshold set at 10, and the read length set at 100 (so as to use the <code class="docutils literal notranslate"><span class="pre">database100mers.kmer_distrib</span></code> file inside the <code class="docutils literal notranslate"><span class="pre">minikraken2_v1_8GB</span></code> folder):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">bracken</span> <span class="o">-</span><span class="n">d</span> <span class="n">path</span><span class="o">/</span><span class="n">to</span><span class="o">/</span><span class="n">minikraken2_v1_8GB</span> <span class="o">-</span><span class="n">i</span> <span class="n">sample</span><span class="o">.</span><span class="n">kraken</span><span class="o">.</span><span class="n">report</span> <span class="o">-</span><span class="n">o</span> <span class="n">sample</span><span class="o">.</span><span class="n">bracken</span> <span class="o">-</span><span class="n">r</span> <span class="mi">100</span> <span class="o">-</span><span class="n">t</span> <span class="mi">10</span>
</pre></div>
</div>
<p>To speed the process, we can also use variables to set up the options and the paths to the minikraken database and other folders, and a <code class="docutils literal notranslate"><span class="pre">for</span></code> loop. We will generate taxonomic abundances at the species level (the default option).
Remember to create first an output folder (with <code class="docutils literal notranslate"><span class="pre">mkdir</span></code>). Also note that you can set as path variables either relative paths (based on your actual position and path) or absolute paths:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>KRAKEN_DB=/path/to/kraken/db-folder
OUTPUT=/path/to/output-folder
READ_LEN=100
THRESHOLD=0

for i in $(find -name &quot;*krk.report&quot; -type f)
do
  FILENAME=$(basename &quot;$i&quot;)
  SAMPLE=${FILENAME%.krk.report}
  bracken -d $KRAKEN_DB -i $i -o $OUTPUT/${SAMPLE}.bracken -r $READ_LEN -t $THRESHOLD
done
</pre></div>
</div>
<p>The Bracken output file is tab-delimited. There are seven fields, from left to right:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>Name</p></li>
<li><p>Taxonomy ID</p></li>
<li><p>Level ID (S=Species, G=Genus, O=Order, F=Family, P=Phylum, K=Kingdom)</p></li>
<li><p>Kraken Assigned Reads</p></li>
<li><p>Added Reads with Abundance Reestimation</p></li>
<li><p>Total Reads after Abundance Reestimation</p></li>
<li><p>Fraction of Total Reads</p></li>
</ol>
</div></blockquote>
</div>
<div class="section" id="bracken-from-custom-databases">
<h3><span class="section-number">3.2.2. </span>Bracken from Custom databases<a class="headerlink" href="#bracken-from-custom-databases" title="Permalink to this headline">¶</a></h3>
<p>The Minikraken database is available with all the files needed to run Bracken. If you used a custom database you must generate yourself the <code class="docutils literal notranslate"><span class="pre">kmer_distrib</span></code> file and the other Bracken database files with <code class="docutils literal notranslate"><span class="pre">bracken-build</span></code>.
To do that, we need the <strong>library</strong> and <strong>taxonomy</strong> of the kraken database that we used for classification. These are big size folders, but do not delete them if you plan to generate Braken databases in the future at different “read-length kmers”.
You can select the “read-length kmers” that is most suitable for your samples (namely the average insert size of your library, e.g. 60 bp for ancient libraries).
You can also generate several <code class="docutils literal notranslate"><span class="pre">kmer_distrib</span></code> files and make different runs (as in Minikraken run, where 100, 150 and 200 are available).
You can find all the steps for generating the Bracken database files in the <a class="reference external" href="https://ccb.jhu.edu/software/bracken/index.shtml?t=manual#step0">Bracken manual</a>.</p>
<blockquote>
<div></div></blockquote>
</div>
</div>
<div class="section" id="mataphlan-2">
<h2><span class="section-number">3.3. </span>Mataphlan 2<a class="headerlink" href="#mataphlan-2" title="Permalink to this headline">¶</a></h2>
<p>MetaPhlAn is a computational tool that relies on ~1.1M unique clade-specific <a class="reference external" href="https://www.dropbox.com/sh/7qze7m7g9fe2xjg/AAAlyQITZuUCtBUJxpxhIroIa/mpa_v30_CHOCOPhlAn_201901_marker_info.txt.bz2?dl=1">marker genes</a> identified from ~100,000 reference genomes (~99,500 bacterial and archaeal and ~500 eukaryotic) to conduct taxonomic profiling of microbial communities (Bacteria, Archaea and Eukaryotes) from metagenomic shotgun sequencing data. MetaPhlAn allows:</p>
<blockquote>
<div><ul class="simple">
<li><p>unambiguous taxonomic assignments;</p></li>
<li><p>an accurate estimation of organismal relative abundance;</p></li>
<li><p>species-level resolution for bacteria, archaea, eukaryotes, and viruses;</p></li>
<li><p>strain identification and tracking</p></li>
<li><p>orders of magnitude speedups compared to existing methods.</p></li>
<li><p>metagenomic strain-level population genomics</p></li>
</ul>
</div></blockquote>
<p>We are not going to cover to in detail MetaPhlAn, but this is a great tool given its specificity, in particular to confirm the detection of peculiar microbial species (e.g. pathogens).
The basic usage of MetaPhlAn for taxonomic profiling is:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">metaphlan2</span><span class="o">.</span><span class="n">py</span> <span class="n">filename</span><span class="o">.</span><span class="n">fastq</span><span class="p">(</span><span class="o">.</span><span class="n">gz</span><span class="p">)</span> <span class="o">--</span><span class="n">input_type</span> <span class="n">fastq</span> <span class="o">&gt;</span> <span class="n">outfile</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>MetaPhlAn relies on BowTie2 to map reads against marker genes. To save the intermediate BowTie2 output use <code class="docutils literal notranslate"><span class="pre">--bowtie2out</span></code>, and for multiple CPUs (if available) use <code class="docutils literal notranslate"><span class="pre">--nproc</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">metaphlan2</span><span class="o">.</span><span class="n">py</span> <span class="n">filename</span><span class="o">.</span><span class="n">fastq</span><span class="p">(</span><span class="o">.</span><span class="n">gz</span><span class="p">)</span> <span class="o">--</span><span class="n">bowtie2out</span> <span class="n">filename</span><span class="o">.</span><span class="n">bowtie2</span><span class="o">.</span><span class="n">bz2</span> <span class="o">--</span><span class="n">nproc</span> <span class="mi">5</span> <span class="o">--</span><span class="n">input_type</span> <span class="n">fastq</span> <span class="o">-</span><span class="n">o</span> <span class="n">output</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>The intermediate BowTie2 output files can be used to run MetaPhlAn quickly by specifying the input (<code class="docutils literal notranslate"><span class="pre">--input_type</span> <span class="pre">bowtie2out</span></code>):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">metaphlan</span> <span class="n">filename</span><span class="o">.</span><span class="n">bowtie2</span><span class="o">.</span><span class="n">bz2</span> <span class="o">--</span><span class="n">nproc</span> <span class="mi">5</span> <span class="o">--</span><span class="n">input_type</span> <span class="n">bowtie2out</span> <span class="o">-</span><span class="n">o</span> <span class="n">output</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
<p>For more information and advanced usage of MetaPhlAn see the <a class="reference external" href="https://github.com/biobakery/MetaPhlAn">manual</a> and the <a class="reference external" href="https://github.com/biobakery/biobakery/wiki/metaphlan2">wiki</a> page. And, recently, <a class="reference external" href="https://github.com/biobakery/MetaPhlAn/wiki/MetaPhlAn-3.0">MetaPhlAn 3</a> has been released!</p>
<blockquote>
<div></div></blockquote>
</div>
</div>
<div class="section" id="other-tools">
<h2><span class="section-number">3.4. </span>Other tools<a class="headerlink" href="#other-tools" title="Permalink to this headline">¶</a></h2>
<p>We remind you that there are several metagenomic classifiers, all with their pros and cons, as discussed in the lecture.
Other tools that we want to suggest are MALT (and the <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1903-0">HOPS</a> pipeline to analyse the data), and <a class="reference external" href="https://ccb.jhu.edu/software/centrifuge/manual.shtml">Centrifuge</a>, which both implement alignment-based methods to classify the reads.</p>
<blockquote>
<div></div></blockquote>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="4_ReadsMapping_v2.html" class="btn btn-neutral float-right" title="4. Alignment of reads to a reference genome" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="2_ReadsFiltering.html" class="btn btn-neutral float-left" title="2. Quality filtering of reads" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Amine Namouchi &amp; Claudio Ottoni

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
    <!-- Theme Analytics -->
    <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-XXXXXXX-1', 'auto');
    ga('send', 'pageview');
    </script>

    
   

</body>
</html>