

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>4. Quality filtering of reads &mdash; Paleogenomics at Tor Vergata 0.1.0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="5. Metagenomic screening of shotgun data" href="4_Metagenomics_v2.html" />
    <link rel="prev" title="3. Working on HPC with Slurm" href="3_Slurm_HPC.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> Paleogenomics at Tor Vergata
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="1_ListTools.html">1. List of Tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="2_Useful_commands.html">2. Useful commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="3_Slurm_HPC.html">3. Working on HPC with Slurm</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">4. Quality filtering of reads</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#reads-quality-control">4.1. Reads quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="#reads-quality-filtering">4.2. Reads quality filtering</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="4_Metagenomics_v2.html">5. Metagenomic screening of shotgun data</a></li>
<li class="toctree-l1"><a class="reference internal" href="5_ReadsMapping_v2.html">6. Alignment of reads to a reference genome</a></li>
<li class="toctree-l1"><a class="reference internal" href="6_VariantsCall.html">7. Variant calling and visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="7_Filtering_SNPs.html">8. Filtering, annotating and combining SNPs</a></li>
<li class="toctree-l1"><a class="reference internal" href="8_HandsOn.html">9. Hands-on</a></li>
<li class="toctree-l1"><a class="reference internal" href="9_old_material.html">10. Material from previous courses</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Paleogenomics at Tor Vergata</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li><span class="section-number">4. </span>Quality filtering of reads</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/3_ReadsFiltering.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="quality-filtering-of-reads">
<h1><span class="section-number">4. </span>Quality filtering of reads<a class="headerlink" href="#quality-filtering-of-reads" title="Permalink to this headline">¶</a></h1>
<div class="section" id="reads-quality-control">
<h2><span class="section-number">4.1. </span>Reads quality control<a class="headerlink" href="#reads-quality-control" title="Permalink to this headline">¶</a></h2>
<p>The reads generated by a sequencing platform are delivered in <code class="docutils literal notranslate"><span class="pre">fastq</span></code> format. The first step of the analysis of raw sequences is the quality-control.
To do that, we will use <a class="reference external" href="https://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>, which provides a modular set of analyses that you can use to have a first impression of whether your data has any problems of which you should be aware before doing any further analysis.
To run FastQC type the following command:</p>
<blockquote>
<div></div></blockquote>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">fastqc</span> <span class="n">filename</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>To analyze multiple <code class="docutils literal notranslate"><span class="pre">fastq</span></code> files you can run <code class="docutils literal notranslate"><span class="pre">FastQC</span></code> as follows:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">fastqc</span> <span class="o">*.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<p>At the end of the analysis, FastQC generates for each input file a summary report, like in the screenshot below:</p>
<img alt="_images/fastqc.png" src="_images/fastqc.png" />
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul>
<li><p>You can download the reports of FastQC (and any other file) in your laptop with the command <code class="docutils literal notranslate"><span class="pre">scp</span></code> (Secure Copy), which allows files to be copied to, from, or between different hosts. It uses ssh for data transfer and provides the same authentication and same level of security as ssh. For example, to copy from a remote host (our server) to your computer:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">scp</span> <span class="n">username</span><span class="nd">@remotehost</span><span class="p">:</span><span class="o">/</span><span class="n">full_path_to_file</span> <span class="o">/</span><span class="n">some</span><span class="o">/</span><span class="n">local</span><span class="o">/</span><span class="n">directory</span>
</pre></div>
</div>
</li>
<li><p>To copy a folder you need to call the option <code class="docutils literal notranslate"><span class="pre">-r</span></code></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">scp</span> <span class="o">-</span><span class="n">r</span> <span class="n">username</span><span class="nd">@remotehost</span><span class="p">:</span><span class="o">/</span><span class="n">full_path_to_file</span> <span class="o">/</span><span class="n">some</span><span class="o">/</span><span class="n">local</span><span class="o">/</span><span class="n">directory</span>
</pre></div>
</div>
</li>
<li><p>If you are using a <code class="docutils literal notranslate"><span class="pre">pem</span></code> file to connect to the server, you have to use in order to download the files:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">scp</span> <span class="o">-</span><span class="n">i</span> <span class="n">filename</span><span class="o">.</span><span class="n">pem</span> <span class="o">-</span><span class="n">r</span> <span class="n">username</span><span class="nd">@remotehost</span><span class="p">:</span><span class="o">/</span><span class="n">full_path_to_file</span> <span class="o">/</span><span class="n">some</span><span class="o">/</span><span class="n">local</span><span class="o">/</span><span class="n">directory</span>
</pre></div>
</div>
</li>
</ul>
</div>
</div>
<div class="section" id="reads-quality-filtering">
<h2><span class="section-number">4.2. </span>Reads quality filtering<a class="headerlink" href="#reads-quality-filtering" title="Permalink to this headline">¶</a></h2>
<p>Reads filtering is a crucial step as it will affect all downstream analyses. One of the important things to do is to trim the adapters that were used during the preparation of the genomic libraries.
For this step we will use the program <a class="reference external" href="https://github.com/MikkelSchubert/adapterremoval">AdapterRemoval</a>, which performs adapter trimming of sequencing reads and subsequent merging (collapse) of paired-end reads with negative insert sizes (an overlap between two sequencing reads derived from a single DNA fragment) into a single collapsed read.
We will analyse the DNA sequences of an ancient sample sequenced in <strong>single-end</strong>. With this command, the trimmed reads are written to <code class="docutils literal notranslate"><span class="pre">output_single.truncated.gz</span></code>, the discarded FASTQ reads are written to <code class="docutils literal notranslate"><span class="pre">output_single.discarded.gz</span></code>, and settings and summary statistics are written to <code class="docutils literal notranslate"><span class="pre">output_single.settings</span></code>.</p>
<blockquote>
<div></div></blockquote>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">AdapterRemoval</span> <span class="o">--</span><span class="n">file1</span> <span class="n">filename</span><span class="o">.</span><span class="n">fastq</span><span class="o">.</span><span class="n">gz</span> <span class="o">--</span><span class="n">basename</span> <span class="n">filename</span> <span class="o">--</span><span class="n">minlength</span> <span class="mi">30</span> <span class="o">--</span><span class="n">trimns</span> <span class="o">--</span><span class="n">trimqualities</span> <span class="o">--</span><span class="n">gzip</span>
</pre></div>
</div>
<p>Here some of the options of AdapterRemoval:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 13%" />
<col style="width: 87%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Option</p></th>
<th class="head"><p>Function</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>-file1</strong> <em>string</em></p></td>
<td><p>Forward reads input file(s) in fastq(.gz) file format. Required option (single-end reads).</p></td>
</tr>
<tr class="row-odd"><td><p><strong>-file2</strong> <em>string</em></p></td>
<td><p>Reverse reads input file(s) in fastq(.gz) file format.</p></td>
</tr>
<tr class="row-even"><td><p><strong>–basename</strong> <em>string</em></p></td>
<td><p>Default prefix for all output files for which no filename was explicitly set [current: your_output]</p></td>
</tr>
<tr class="row-odd"><td><p><strong>–adapter1</strong> <em>sequence</em></p></td>
<td><p>Adapter sequence expected to be found in mate 1 reads [current: AGATCGGAAGAGCACACGTCTGAACTCCAGTCACNNNNNNATCTCGTATGCCGTCTTCTGCTTG]</p></td>
</tr>
<tr class="row-even"><td><p><strong>–adapter2</strong> <em>sequence</em></p></td>
<td><p>Adapter sequence expected to be found in mate 2 reads [current: AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGTAGATCTCGGTGGTCGCCGTATCATT]</p></td>
</tr>
<tr class="row-odd"><td><p><strong>–trimns</strong></p></td>
<td><p>If set, trim ambiguous bases (N) at 5’/3’ termini [current: off]</p></td>
</tr>
<tr class="row-even"><td><p><strong>–trimqualities</strong></p></td>
<td><p>If set, trim bases at 5’/3’ termini with quality scores &lt;= to <strong>–minquality</strong> value [current: off]</p></td>
</tr>
<tr class="row-odd"><td><p><strong>–minquality</strong> <em>integer</em></p></td>
<td><p>PHRED inclusive minimum values; see <strong>–trimqualities</strong> for details [current: 2]</p></td>
</tr>
<tr class="row-even"><td><p><strong>–minlength</strong> <em>integer</em></p></td>
<td><p>Reads shorter than this length are discarded following trimming [current: 15].</p></td>
</tr>
<tr class="row-odd"><td><p><strong>–collapse</strong></p></td>
<td><p>When set, paired ended read alignments of <strong>–minalignmentlength</strong> or more bases are combined into a single consensus sequence, representing the complete insert</p></td>
</tr>
<tr class="row-even"><td><p><strong>–minalignmentlength</strong> <em>integer</em></p></td>
<td><p>If <strong>–collapse</strong> is set, paired reads must overlap at least this number of bases to be collapsed, and single-ended reads must overlap at least this number of bases with the adapter to be considered complete template molecules [current: 11].</p></td>
</tr>
</tbody>
</table>
<div class="admonition note">
<p class="admonition-title">Note</p>
<ul>
<li><p>Here we are using a single-end library, for <strong>paired-end</strong> libraries use the following command. The merged reads are in the file <code class="docutils literal notranslate"><span class="pre">output.collapsed.gz</span></code>. The <code class="docutils literal notranslate"><span class="pre">output.collapsed.truncated.gz</span></code> contains merged reads that have been trimmed due to the <strong>–trimns</strong> or <strong>–trimqualities</strong> options. The files <code class="docutils literal notranslate"><span class="pre">output.pair1.truncated.gz</span></code> and <code class="docutils literal notranslate"><span class="pre">output.pair2.truncated.gz</span></code> contain trimmed pairs of reads which were not collapsed, <code class="docutils literal notranslate"><span class="pre">output_paired.singleton.truncated</span></code> contains reads where one mate was discarded.
Finally, the <code class="docutils literal notranslate"><span class="pre">output_paired.discarded</span></code> and <code class="docutils literal notranslate"><span class="pre">output_paired.settings</span></code> files correspond to those of the single-end run.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">AdapterRemoval</span> <span class="o">--</span><span class="n">file1</span> <span class="n">filename_R1</span><span class="o">.</span><span class="n">fastq</span> <span class="o">--</span><span class="n">file2</span> <span class="n">filename_R2</span><span class="o">.</span><span class="n">fastq</span> <span class="o">--</span><span class="n">basename</span> <span class="n">filename</span> <span class="o">--</span><span class="n">minlength</span> <span class="mi">30</span> <span class="o">--</span><span class="n">trimns</span> <span class="o">--</span><span class="n">trimqualities</span> <span class="o">--</span><span class="n">minquality</span> <span class="mi">25</span> <span class="o">--</span><span class="n">collapse</span> <span class="o">--</span><span class="n">gzip</span>
</pre></div>
</div>
</li>
<li><p>Several tools can be used for reads pre-processing and filtering, for example: <a class="reference external" href="https://github.com/apeltzer/ClipAndMerge">ClipAndMerge</a>, <a class="reference external" href="https://github.com/grenaud/leeHom">leeHom</a>, <a class="reference external" href="https://github.com/jdidion/atropos">Atropos</a>, <a class="reference external" href="https://github.com/OpenGene/fastp">fastp</a>.</p>
<blockquote>
<div></div></blockquote>
</li>
</ul>
</div>
<p>After reads filtering open your adapter-trimmed <code class="docutils literal notranslate"><span class="pre">fastq</span></code> file again in FastQC and see the differences before (the two original paired-end reads files) and after (the collapsed reads file) adapter trimming.</p>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="4_Metagenomics_v2.html" class="btn btn-neutral float-right" title="5. Metagenomic screening of shotgun data" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="3_Slurm_HPC.html" class="btn btn-neutral float-left" title="3. Working on HPC with Slurm" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2021, Claudio Ottoni

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
    <!-- Theme Analytics -->
    <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-XXXXXXX-1', 'auto');
    ga('send', 'pageview');
    </script>

    
   

</body>
</html>