

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>6. Filtering, annotating and combining SNPs &mdash; Physalia Paleogenomics 0.1.0 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  
  
  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script src="_static/language_data.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="7. DO-IT-YOURSELF" href="9_DIY.html" />
    <link rel="prev" title="5. Variant calling and visualization" href="7_VariantsCall.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="index.html" class="icon icon-home" alt="Documentation Home"> Physalia Paleogenomics
          

          
          </a>

          
            
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="1_ListTools.html">1. List of Tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="2_ReadsFiltering.html">2. Quality filtering of reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="3_Metagenomics_v2.html">3. Metagenomic screening of shotgun data</a></li>
<li class="toctree-l1"><a class="reference internal" href="4_ReadsMapping_v2.html">4. Alignment of reads to a reference genome</a></li>
<li class="toctree-l1"><a class="reference internal" href="7_VariantsCall.html">5. Variant calling and visualization</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">6. Filtering, annotating and combining SNPs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#snps-filtering-and-annotion">6.1. SNPs filtering and annotion</a></li>
<li class="toctree-l2"><a class="reference internal" href="#compare-and-combine-multiple-annotation-files">6.2. Compare and combine multiple annotation files</a></li>
<li class="toctree-l2"><a class="reference internal" href="#phylogenetic-tree-reconstruction">6.3. Phylogenetic tree reconstruction</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="9_DIY.html">7. DO-IT-YOURSELF</a></li>
<li class="toctree-l1"><a class="reference internal" href="10_old_material.html">8. Material from previous courses</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Physalia Paleogenomics</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
        
      <li><span class="section-number">6. </span>Filtering, annotating and combining SNPs</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/8_Filtering_SNPs.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="filtering-annotating-and-combining-snps">
<h1><span class="section-number">6. </span>Filtering, annotating and combining SNPs<a class="headerlink" href="#filtering-annotating-and-combining-snps" title="Permalink to this headline">¶</a></h1>
<p>To investigate the genetic variants in the <code class="docutils literal notranslate"><span class="pre">vcf</span></code> files we will use the program <a class="reference external" href="https://github.com/Amine-Namouchi/snpToolkit">snpToolkit</a>.
The <code class="docutils literal notranslate"><span class="pre">-h</span></code> option will display the following message:</p>
<blockquote>
<div></div></blockquote>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ snptoolkit -h
usage: snptoolkit [-h] {explore,annotate,combine,viz,expand} ...

    snpToolkit takes vcf files, as well as bam files (optional) as inputs. The vcf files could be generated using samtools/bcftools, gatk HaplotypeCaller or freeBayes.
    Please visit https://github.com/Amine-Namouchi/snpToolkit for more information.

positional arguments:
  {explore,annotate,combine,viz,expand}
                        commands
    explore             explore your vcf files before annotation
    annotate            Annotate one or multiple vcf files
    combine             combine snpToolkit output files in one alignment in fasta format
    viz                 visualize snptoolkit output files
    expand              expand existent list of polymorphic sites when new SNP output files are availble

optional arguments:
  -h, --help            show this help message and exit
</pre></div>
</div>
<p>Five options are possible: <code class="docutils literal notranslate"><span class="pre">explore</span></code>, <code class="docutils literal notranslate"><span class="pre">annotate</span></code>, <code class="docutils literal notranslate"><span class="pre">combine</span></code>, <code class="docutils literal notranslate"><span class="pre">viz</span></code>, <code class="docutils literal notranslate"><span class="pre">expand</span></code>.</p>
<div class="section" id="snps-filtering-and-annotion">
<h2><span class="section-number">6.1. </span>SNPs filtering and annotion<a class="headerlink" href="#snps-filtering-and-annotion" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal notranslate"><span class="pre">snpToolkit</span> <span class="pre">annotate</span></code> command will display general information about the usage of the program:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>snptoolkit annotate -h
usage: snptoolkit annotate <span class="o">[</span>-h<span class="o">]</span> -i IDENTIFIER -g GENBANK <span class="o">[</span>-p PROCESSORS<span class="o">]</span> <span class="o">[</span>-f EXCLUDECLOSESNPS<span class="o">]</span> <span class="o">[</span>-q QUALITY<span class="o">]</span> <span class="o">[</span>-d DEPTH<span class="o">]</span> <span class="o">[</span>-r RATIO<span class="o">]</span> <span class="o">[</span>-e EXCLUDE<span class="o">]</span>

optional arguments:
  -h, --help           show this <span class="nb">help</span> message and <span class="nb">exit</span>

snpToolkit annotate required options:
  -i IDENTIFIER        provide a specific identifier to recognize the file<span class="o">(</span>s<span class="o">)</span> to be analyzed
  -g GENBANK           Pleae provide a genbank file

snpToolkit annotate additional options:
  -p PROCESSORS        number of vcf files to be annotated in parallel default value <span class="o">[</span><span class="m">1</span><span class="o">]</span>
  -f EXCLUDECLOSESNPS  exclude SNPs <span class="k">if</span> the distance between them is lower <span class="k">then</span> the specified window size in bp
  -q QUALITY           quality score to consider as a cutoff <span class="k">for</span> variant calling. default value <span class="o">[</span><span class="m">20</span><span class="o">]</span>
  -d DEPTH             minimum depth caverage. default value <span class="o">[</span><span class="m">3</span><span class="o">]</span>
  -r RATIO             minimum ratio that correspond to the number of reads that has the mutated allele / total depth in that particular position. default
                      value <span class="o">[</span><span class="m">0</span><span class="o">]</span>
  -e EXCLUDE           provide a tab file with genomic regions to exclude in this format: region start stop. region must correspond to the same name<span class="o">(</span>s<span class="o">)</span> of
                      chromsome and plasmids as in the genbank file
</pre></div>
</div>
<p>Here is a simple example on how to use snpToolkit:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">snptoolkit</span> <span class="n">annotate</span> <span class="o">-</span><span class="n">i</span> <span class="n">vcf</span> <span class="o">-</span><span class="n">g</span> <span class="n">GCF_000009065</span><span class="o">.</span><span class="mi">1</span><span class="n">_ASM906v1_genomic</span><span class="o">.</span><span class="n">gbff</span> <span class="o">-</span><span class="n">d</span> <span class="mi">3</span> <span class="o">-</span><span class="n">q</span> <span class="mi">20</span> <span class="o">-</span><span class="n">r</span> <span class="mf">0.9</span>
</pre></div>
</div>
<p>snpToolkit can automatically recogninze <code class="docutils literal notranslate"><span class="pre">vcf</span></code> files generated with the following programs: <code class="docutils literal notranslate"><span class="pre">samtools</span> <span class="pre">mpileup</span></code>, <code class="docutils literal notranslate"><span class="pre">gatk</span> <span class="pre">HaplotyCaller</span></code> and <code class="docutils literal notranslate"><span class="pre">freeBayes</span></code>. The <code class="docutils literal notranslate"><span class="pre">vcf</span></code> files could be gzipped or not. In the command line above, snpToolkit will filter and annotate all SNPs in the <code class="docutils literal notranslate"><span class="pre">vcf</span></code> file(s) that fullfil the following criteria: <code class="docutils literal notranslate"><span class="pre">quality</span> <span class="pre">&gt;=</span> <span class="pre">30</span></code>, <code class="docutils literal notranslate"><span class="pre">depth</span> <span class="pre">of</span> <span class="pre">coverage</span> <span class="pre">&gt;=</span> <span class="pre">5</span></code> and <code class="docutils literal notranslate"><span class="pre">ratio</span> <span class="pre">&gt;=</span> <span class="pre">0.9</span></code>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For each SNP position, the ratio (r) is calculated as follows:</p>
<p>r= dm / (dr + dm)</p>
<ul class="simple">
<li><p>dr= Number of reads having the reference allele</p></li>
<li><p>dm= Number of reads having the mutated allele</p></li>
</ul>
</div>
<p>The output file(s) of snpToolkit is a tabulated file(s) that you can open with Microsoft Excel and it will look as follow:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1">##snpToolkit=__version__</span>
<span class="c1">##commandline= snptoolkit annotate -i vcf -g GCF_000009065.1_ASM906v1_genomic.gbff -d 5 -q 30 -r 0.9 -p 4</span>
<span class="c1">##VcfFile=sample5.vcf.gz</span>
<span class="c1">##Total number of SNPs before snpToolkit processing: 406</span>
<span class="c1">##The options -f and -e were not used</span>
<span class="c1">##Filtred SNPs. Among the 406 SNPs, the number of those with a quality score &gt;= 30, a depth &gt;= 5 and a ratio &gt;= 0.9 is: 218</span>
<span class="c1">##After mapping, SNPs were located in:</span>
<span class="c1">##NC_003131.1: Yersinia pestis CO92 plasmid pCD1, complete sequence 70305 bp</span>
<span class="c1">##NC_003143.1: Yersinia pestis CO92, complete genome 4653728 bp</span>
<span class="c1">##The mapped and annotated SNPs are distributed as follow:</span>
<span class="c1">##Location      Genes   RBS     tRNA    rRNA    ncRNA   Pseudogenes     intergenic      Synonymous      NonSynonumous</span>
<span class="c1">##SNPs in NC_003143.1: Yersinia pestis CO92, complete genome 4653728 bp 155     0       0       1       0       0       57      54      101</span>
<span class="c1">##SNPs in NC_003131.1: Yersinia pestis CO92 plasmid pCD1, complete sequence 70305 bp    2       0       0       0       0       0       3       1       1</span>
<span class="c1">##Syn=Synonymous NS=Non-Synonymous</span>
<span class="c1">##Coordinates   REF     SNP     Depth   Nb of reads REF Nb reads SNPs   Ratio   Quality Annotation      Product Orientation     Coordinates in gene     Ref codon       SNP codon       Ref AA  SNP AA  Coordinates protein     Effect  Location</span>
<span class="m">82</span>      C       A       <span class="m">36</span>      <span class="m">0</span>       <span class="m">34</span>      <span class="m">1</span>.0     <span class="m">138</span>.0   intergenic      .       +       .       -       -       -       -       -       -       NC_003143.1: Yersinia pestis CO92, <span class="nb">complete</span> genome <span class="m">4653728</span> bp
<span class="m">130</span>     G       C       <span class="m">28</span>      <span class="m">0</span>       <span class="m">27</span>      <span class="m">1</span>.0     <span class="m">144</span>.0   intergenic      .       +       .       -       -       -       -       -       -       NC_003143.1: Yersinia pestis CO92, <span class="nb">complete</span> genome <span class="m">4653728</span> bp
<span class="m">855</span>     G       A       <span class="m">69</span>      <span class="m">0</span>       <span class="m">62</span>      <span class="m">1</span>.0     <span class="m">228</span>.0   YPO_RS01010<span class="p">|</span>asnC        transcriptional regulator AsnC  -       <span class="m">411</span>     ACC     AC<span class="o">[</span>T<span class="o">]</span>   T       T       <span class="m">137</span>     Syn     NC_003143.1: Yersinia pestis CO92, <span class="nb">complete</span> genome <span class="m">4653728</span> bp
</pre></div>
</div>
<p>The header of the generated snpToolkit output file includes useful information e.g. raw number of SNPs, Number of filtered SNPs, SNPs distribution, etc…
The SNPs annotation is organized in tab delimited table. The columns of this table are:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 32%" />
<col style="width: 68%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Column name</p></th>
<th class="head"><p>Description</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Coordinates</p></td>
<td><p>SNP coordinate</p></td>
</tr>
<tr class="row-odd"><td><p>REF</p></td>
<td><p>Reference allele</p></td>
</tr>
<tr class="row-even"><td><p>SNP</p></td>
<td><p>New allele in analyzed sample</p></td>
</tr>
<tr class="row-odd"><td><p>Depth</p></td>
<td><p>Total depth of coverage</p></td>
</tr>
<tr class="row-even"><td><p>Nb of reads REF</p></td>
<td><p>Number of reads with the reference allele</p></td>
</tr>
<tr class="row-odd"><td><p>Nb reads SNPs</p></td>
<td><p>Number of reads with the new allele</p></td>
</tr>
<tr class="row-even"><td><p>Ratio</p></td>
<td><p>Nb reads SNPs/(Nb of reads REF+Nb reads SNPs)</p></td>
</tr>
<tr class="row-odd"><td><p>Quality</p></td>
<td><p>Quality score</p></td>
</tr>
<tr class="row-even"><td><p>Annotation</p></td>
<td><p>Distribution within genes or intergenic</p></td>
</tr>
<tr class="row-odd"><td><p>Product</p></td>
<td><p>Functional product of the gene</p></td>
</tr>
<tr class="row-even"><td><p>Orientation</p></td>
<td><p>Gene orientation</p></td>
</tr>
<tr class="row-odd"><td><p>Coordinates in gene</p></td>
<td><p>Coordinate of the SNP within the gene</p></td>
</tr>
<tr class="row-even"><td><p>Ref codon</p></td>
<td><p>Reference codon, ACC in the example above</p></td>
</tr>
<tr class="row-odd"><td><p>SNP codon</p></td>
<td><p>New codon, AC[T]</p></td>
</tr>
<tr class="row-even"><td><p>Ref AA</p></td>
<td><p>Amino Acid corresponding to reference codon</p></td>
</tr>
<tr class="row-odd"><td><p>SNP AA</p></td>
<td><p>Amino Acid corresponding to new codon</p></td>
</tr>
<tr class="row-even"><td><p>Coordinates protein</p></td>
<td><p>Coordinate of the Amino Acid</p></td>
</tr>
<tr class="row-odd"><td><p>Effect</p></td>
<td><p>Could be Synonymous (Syn) or Non-Synonymous (NS)</p></td>
</tr>
<tr class="row-even"><td><p>Location</p></td>
<td><p>ID of the chromosome and plasmids.</p></td>
</tr>
</tbody>
</table>
</div>
<div class="section" id="compare-and-combine-multiple-annotation-files">
<h2><span class="section-number">6.2. </span>Compare and combine multiple annotation files<a class="headerlink" href="#compare-and-combine-multiple-annotation-files" title="Permalink to this headline">¶</a></h2>
<p>After generating a set of output files, you can run <code class="docutils literal notranslate"><span class="pre">snpToolkit</span> <span class="pre">combine</span></code>:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>$ snptoolkit combine  -h
usage: snptoolkit combine [-h] --location LOCATION [-r RATIO] [--bam BAMFILTER BAMFILTER BAMFILTER] [--snps {ns,s,all,inter}] [-e EXCLUDE]

optional arguments:
  -h, --help            show this help message and exit

snpToolkit combine required options:
  --location LOCATION   provide for example the name of the chromosome or plasmid you want to create fasta alignemnt for

snpToolkit additional options:
  -r RATIO              new versus reference allele ratio to filter SNPs from snpToolkit outputs. default [0]
  --bam BAMFILTER BAMFILTER BAMFILTER
                        provide the depth, ratio and the path to the folder containing the bam files. eg. 3 0.9 path
  --snps {ns,s,all,inter}
                        Specify if you want to concatenate all SNPs or just synonymous (s), non-synonymous (ns) or intergenic (inter) SNPs. default [all]
  -e EXCLUDE            Provide a yaml file with keywords and coordinates to be excluded
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">snpToolkit</span> <span class="pre">combine</span></code> will compare all the SNPs identified in each file and create two additional output files:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>a tabulated files with all polymorphic sites</p></li>
<li><p>a <code class="docutils literal notranslate"><span class="pre">fasta</span></code> file.</p></li>
</ol>
</div></blockquote>
<p>To combine the snps from different samples in one alignment <code class="docutils literal notranslate"><span class="pre">fasta</span></code> file you type the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">snptoolkit</span> <span class="n">combine</span> <span class="o">--</span><span class="n">loc</span> <span class="n">NC_003143</span><span class="o">.</span><span class="mi">1</span> <span class="o">-</span><span class="n">r</span> <span class="mf">0.9</span> <span class="o">--</span><span class="n">bam</span> <span class="mi">2</span> <span class="mf">1.0</span> <span class="o">../</span><span class="n">bam</span><span class="o">/</span>
</pre></div>
</div>
<p>As we will be working with ancient DNA, a small fraction of your genome could be covered. In this case we will use the option <code class="docutils literal notranslate"><span class="pre">--bam</span></code> to indicate the path to the folder containing the <code class="docutils literal notranslate"><span class="pre">bam</span></code> files.
The option <code class="docutils literal notranslate"><span class="pre">-d</span></code> must be used with the option <code class="docutils literal notranslate"><span class="pre">--bam</span></code>. By default, all SNPs will be reported. This behaviour can be changed using the option <code class="docutils literal notranslate"><span class="pre">--snp</span></code>.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>It is also possible to use the option <code class="docutils literal notranslate"><span class="pre">--bam</span></code> with modern data as some genomic regions could be deleted.</p>
</div>
<p>The file reporting the polymorphic sites is organized as follows:</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 5%" />
<col style="width: 14%" />
<col style="width: 4%" />
<col style="width: 4%" />
<col style="width: 36%" />
<col style="width: 9%" />
<col style="width: 9%" />
<col style="width: 9%" />
<col style="width: 9%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>ID</p></th>
<th class="head"><p>Coordinates</p></th>
<th class="head"><p>REF</p></th>
<th class="head"><p>SNP</p></th>
<th class="head"><p>Columns with SNP information</p></th>
<th class="head"><p>sample1</p></th>
<th class="head"><p>sample2</p></th>
<th class="head"><p>sample3</p></th>
<th class="head"><p>sample4</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>snp1</p></td>
<td><p>130</p></td>
<td><p>A</p></td>
<td><p>T</p></td>
<td></td>
<td><p>1</p></td>
<td><p>1</p></td>
<td><p>1</p></td>
<td><p>1</p></td>
</tr>
<tr class="row-odd"><td><p>snp2</p></td>
<td><p>855</p></td>
<td><p>C</p></td>
<td><p>G</p></td>
<td></td>
<td><p>0</p></td>
<td><p>0</p></td>
<td><p>?</p></td>
<td><p>1</p></td>
</tr>
<tr class="row-even"><td><p>snp3</p></td>
<td><p>1315</p></td>
<td><p>A</p></td>
<td><p>C</p></td>
<td></td>
<td><p>1</p></td>
<td><p>1</p></td>
<td><p>0</p></td>
<td><p>0</p></td>
</tr>
<tr class="row-odd"><td><p>snp4</p></td>
<td><p>12086</p></td>
<td><p>G</p></td>
<td><p>A</p></td>
<td></td>
<td><p>1</p></td>
<td><p>0</p></td>
<td><p>?</p></td>
<td><p>0</p></td>
</tr>
</tbody>
</table>
<p>The table above reports the distribution of all polymorphic sites in all provided files.
As we provided the <code class="docutils literal notranslate"><span class="pre">bam</span></code> files of the ancient DNA samples, snpToolkit will check if the polymorphic sites (snp2 and snp4) are absent in sample3
because there is no SNP in that positions or because the region where the snps are located is not covered. In the latter case, snpToolkit will add a question mark <code class="docutils literal notranslate"><span class="pre">?</span></code> that reflects a missing data.
From the table above, it will be possible to generate a <code class="docutils literal notranslate"><span class="pre">fasta</span></code> file, like the one below:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>&gt;Reference
ATCGGGTATGCCAATGCGT
&gt;Sample1
ACCGGGTATGCCAATGTGT
&gt;Sample2
ATTGGGTATGCCAGTGCGT
&gt;Sample3
?TTGAGT?TGTCA?TACGT
&gt;Sample4
ATCGGGTATGCCAATGCGT
</pre></div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">fasta</span></code> output file will be used to generate a maximum likelihood tree using <code class="docutils literal notranslate"><span class="pre">IQ-TREE</span></code></p>
</div>
<div class="section" id="phylogenetic-tree-reconstruction">
<h2><span class="section-number">6.3. </span>Phylogenetic tree reconstruction<a class="headerlink" href="#phylogenetic-tree-reconstruction" title="Permalink to this headline">¶</a></h2>
<p>There are several tools to build phylogenetic trees. All of these tools, use an alignment file as input file. Now that we have generated an alignment file in <code class="docutils literal notranslate"><span class="pre">fasta</span></code> format, we will use <code class="docutils literal notranslate"><span class="pre">IQ-TREE</span></code> to build a maximum likelihood tree.
We use <code class="docutils literal notranslate"><span class="pre">IQ-TREE</span></code> for several reasons:</p>
<ul class="simple">
<li><p>It performs a composition chi-square test for every sequence in the alignment. A sequence is denoted failed if its character composition significantly deviates from the average composition of the alignment.</p></li>
<li><p>Availability of a wide variety of phylogenetic models. <code class="docutils literal notranslate"><span class="pre">IQ-TREE</span></code> uses <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pubmed/28481363">ModelFinder</a> to find the best substitution model that will be used directly to build the maximum likelihood phylogenetic tree.</p></li>
<li><p>Multithreading</p></li>
</ul>
<p>To generate the phylogenetic tree type the following command using your <code class="docutils literal notranslate"><span class="pre">fasta</span></code> as input:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">iqtree</span> <span class="o">-</span><span class="n">m</span> <span class="n">MFP</span><span class="o">+</span><span class="n">ASC</span> <span class="o">-</span><span class="n">s</span> <span class="n">SNPs_alignment</span><span class="o">.</span><span class="n">fasta</span>
</pre></div>
</div>
<table class="docutils align-default">
<colgroup>
<col style="width: 27%" />
<col style="width: 73%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>IQ-TREE options</p></th>
<th class="head"><p>Function</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><strong>-m</strong></p></td>
<td><p><a class="reference external" href="http://www.iqtree.org/doc/Substitution-Models">Substitution</a> model name to use during the analysis.</p></td>
</tr>
<tr class="row-odd"><td><p><strong>-s</strong></p></td>
<td><p>Alignment file</p></td>
</tr>
</tbody>
</table>
<p>More information of IQ-TREE can be found in the program’s <a class="reference external" href="http://www.iqtree.org/doc/Tutorial">tutorial</a></p>
<blockquote>
<div></div></blockquote>
<p>The phylogenetic tree generated can be visualized using <code class="docutils literal notranslate"><span class="pre">Figtree</span></code>, <a class="reference external" href="http://tree.bio.ed.ac.uk/software/figtree">download</a> it in your local machine and load the <code class="docutils literal notranslate"><span class="pre">treefile</span></code> output from IQ-TREE to visualize the tree.</p>
<blockquote>
<div></div></blockquote>
</div>
</div>


           </div>
           
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="9_DIY.html" class="btn btn-neutral float-right" title="7. DO-IT-YOURSELF" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="7_VariantsCall.html" class="btn btn-neutral float-left" title="5. Variant calling and visualization" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        
        &copy; Copyright 2020, Amine Namouchi &amp; Claudio Ottoni

    </p>
  </div>
    
    
    
    Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/rtfd/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
    <!-- Theme Analytics -->
    <script>
    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-XXXXXXX-1', 'auto');
    ga('send', 'pageview');
    </script>

    
   

</body>
</html>