<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>10. Analysis of Metagenomic data &mdash; Paleogenomics at Tor Vergata 0.1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
        <script async="async" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="prev" title="9. Filtering, annotating and combining SNPs" href="7_Filtering_SNPs.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> Paleogenomics at Tor Vergata
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="1_ListTools.html">1. List of Tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="2_Terminal_setup.html">2. First step to set-up your machine</a></li>
<li class="toctree-l1"><a class="reference internal" href="2_Useful_commands.html">3. Useful commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="3_Slurm_HPC.html">4. Working on HPC with Slurm</a></li>
<li class="toctree-l1"><a class="reference internal" href="3_ReadsFiltering.html">5. Quality filtering of reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="4_Metagenomics_v2.html">6. Metagenomic screening of shotgun data</a></li>
<li class="toctree-l1"><a class="reference internal" href="5_ReadsMapping_v2.html">7. Alignment of reads to a reference genome</a></li>
<li class="toctree-l1"><a class="reference internal" href="6_VariantsCall.html">8. Variant calling and visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="7_Filtering_SNPs.html">9. Filtering, annotating and combining SNPs</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">10. Analysis of Metagenomic data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#prepare-abundance-tables-for-downstream-analyses">10.1. Prepare abundance tables for downstream analyses</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#parse-the-kraken-bracken-outputs-to-abundance-tables">10.1.1. Parse the kraken/bracken outputs to abundance tables</a></li>
<li class="toctree-l3"><a class="reference internal" href="#normalization-of-the-abundance-table-for-genome-length">10.1.2. Normalization of the abundance table for genome length</a></li>
<li class="toctree-l3"><a class="reference internal" href="#retrieve-full-taxonomic-ranks-of-the-abundance-table">10.1.3. Retrieve full taxonomic ranks of the abundance table</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#sourcetracker">10.2. Sourcetracker</a></li>
<li class="toctree-l2"><a class="reference internal" href="#analyses-with-phyloseq">10.3. Analyses with Phyloseq</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#importing-abundance-tables-in-phyloseq">10.3.1. Importing abundance tables in Phyloseq</a></li>
<li class="toctree-l3"><a class="reference internal" href="#barplot-of-taxa-abundances">10.3.2. Barplot of taxa abundances</a></li>
<li class="toctree-l3"><a class="reference internal" href="#principal-coordinate-analysis-multidimensional-scaling">10.3.3. Principal Coordinate Analysis (Multidimensional Scaling)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#non-metric-multidimensional-scaling">10.3.4. Non-Metric Multidimensional Scaling</a></li>
<li class="toctree-l3"><a class="reference internal" href="#differential-taxonomic-abundances-with-deseq2">10.3.5. Differential taxonomic abundances with DESeq2</a></li>
<li class="toctree-l3"><a class="reference internal" href="#principal-component-analysis-pca">10.3.6. Principal Component Analysis (PCA)</a></li>
</ul>
</li>
</ul>
</li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Paleogenomics at Tor Vergata</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li><span class="section-number">10. </span>Analysis of Metagenomic data</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/9_Metagenomic_analysis.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="analysis-of-metagenomic-data">
<h1><span class="section-number">10. </span>Analysis of Metagenomic data<a class="headerlink" href="#analysis-of-metagenomic-data" title="Permalink to this headline"></a></h1>
<p>Here we analyse shotgun sequence data from dental calculus. After filtering and collapsing the reads, classify the reads of each sample with Kraken2 and estimate the abudances with Bracken.
After that, you can use R to run analyses and create charts from the abundance tables generated with Bracken.</p>
<section id="prepare-abundance-tables-for-downstream-analyses">
<h2><span class="section-number">10.1. </span>Prepare abundance tables for downstream analyses<a class="headerlink" href="#prepare-abundance-tables-for-downstream-analyses" title="Permalink to this headline"></a></h2>
<section id="parse-the-kraken-bracken-outputs-to-abundance-tables">
<h3><span class="section-number">10.1.1. </span>Parse the kraken/bracken outputs to abundance tables<a class="headerlink" href="#parse-the-kraken-bracken-outputs-to-abundance-tables" title="Permalink to this headline"></a></h3>
<p>Run the <code class="docutils literal notranslate"><span class="pre">brackenToAbundanceTable_v2.R</span></code> R script to merge the species aundances of all the samples in one table. The script needs as argument the path to the folder containing the bracken results.
Move to the bracken results folder and type the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">brackenToAbundanceTable_v2</span><span class="o">.</span><span class="n">R</span> <span class="o">.</span>
</pre></div>
</div>
<p>The script will generate the abundance tables, <code class="docutils literal notranslate"><span class="pre">taxa_abundance_bracken_names_IDs.txt</span></code>, which contains the species names and the NCBI IDs.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You can generate abundance tables for various datasets (with bracken files contained in dedicated folders) in different folders and rename the final tables based on the folder (i.e. dataset) name. Try the following commands in the the folder containing all the datasets-subfolders:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>for i in $(find -name &quot;*names_IDs.txt&quot; -type f | sort); do tag=$(basename $(dirname &quot;$i&quot;)); cp $i ${FOLDER}/${tag}_abundance_bracken_names_IDs.txt; done
</pre></div>
</div>
</div>
<p>Then use the <code class="docutils literal notranslate"><span class="pre">abundanceTablesMerger_v2.R</span></code> script to merge all the tables to obtain one comprehensive table (<code class="docutils literal notranslate"><span class="pre">abundance_table_IDs.merged</span></code>) that you will use for downstream analyses. Remember to rename each file based on the specific dataset before pooling them:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">abundanceTablesMerger_v2</span><span class="o">.</span><span class="n">R</span> <span class="o">*.</span><span class="n">txt</span>
</pre></div>
</div>
</section>
<section id="normalization-of-the-abundance-table-for-genome-length">
<h3><span class="section-number">10.1.2. </span>Normalization of the abundance table for genome length<a class="headerlink" href="#normalization-of-the-abundance-table-for-genome-length" title="Permalink to this headline"></a></h3>
<p>Since the Kraken database was built from complete Bacterial, Archaeal and Viral genomes, we must make a normalization for the genome leghts of each species. This normalization is important when we want to analyse relative taxa abundances to characterise full microbiomes.
To do that will use a Python <a class="reference external" href="https://github.com/claottoni/toolbox/tree/main/gL-nomalizer">script</a> that takes three arguments:</p>
<blockquote>
<div></div></blockquote>
<ol class="arabic simple">
<li><p>The abundance table with names</p></li>
<li><p>The table of genome lengths (parsed from the NCBI genome <a class="reference external" href="https://www.ncbi.nlm.nih.gov/genome/browse/#!/prokaryotes/">browser</a>)</p></li>
<li><p>The name of the output file</p></li>
</ol>
<blockquote>
<div></div></blockquote>
<p>The script searches the <strong>names</strong> of the species in the file with the list of genome lengths of species in the NCBI and divides the abundance of each species by the length of the assembly available in the NCBI.
Run the normalization with <code class="docutils literal notranslate"><span class="pre">gL-normalizer-lite_v2.py</span></code> and type the command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">gL</span><span class="o">-</span><span class="n">normalizer</span><span class="o">-</span><span class="n">lite_v2</span><span class="o">.</span><span class="n">py</span> <span class="n">abundance_table_IDs</span><span class="o">.</span><span class="n">merged</span> <span class="n">prokaryotes_viruses_organelles</span><span class="o">.</span><span class="n">table</span> <span class="n">abundance_table_IDs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">norm</span>
</pre></div>
</div>
<p>The output table reports the <strong>Species</strong>, the <strong>NCBI IDs</strong>, the <strong>Assembly stage</strong> of the genome in NCBI, the <strong>Length in Megabases</strong>, the kind of <strong>Match</strong>, exact, species (when a peculiar strain is not found in the list of lenghts), and genus (when a species is not found).</p>
</section>
<section id="retrieve-full-taxonomic-ranks-of-the-abundance-table">
<h3><span class="section-number">10.1.3. </span>Retrieve full taxonomic ranks of the abundance table<a class="headerlink" href="#retrieve-full-taxonomic-ranks-of-the-abundance-table" title="Permalink to this headline"></a></h3>
<p>In the next step, you will retrieve the full taxonomic ranks for the species of your abundance table. To do that you must have the program <a class="reference external" href="https://github.com/linzhi2013/taxonomy_ranks">Taxaranks</a> installed.</p>
<blockquote>
<div></div></blockquote>
<p>Run the following script, which incorporates Taxaranks, to attach the full taxonomy to every species ID.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">getFullTaxaranks</span><span class="o">.</span><span class="n">sh</span> <span class="o">-</span><span class="n">i</span> <span class="n">abundance_table_IDs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">norm</span>
</pre></div>
</div>
<p>The script generates the following files:</p>
<ol class="arabic simple">
<li><p>a <code class="docutils literal notranslate"><span class="pre">file.taxonomy</span></code> - the full taxonomic ranks (from Kingdom) of the species entries.</p></li>
<li><p>a <code class="docutils literal notranslate"><span class="pre">file.taxonomy.err</span></code> - the species for which no taxonomic ranks could be found.</p></li>
<li><p>a <code class="docutils literal notranslate"><span class="pre">file.taxonomy.final</span></code> - full taxonomy ranks are included in the original abundance table.</p></li>
</ol>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <code class="docutils literal notranslate"><span class="pre">file.taxonomy.err</span></code> should be empty in principle. If not, then the ful taxonomic ranks could not be retrieved for the species listed. In that case, try to update the NCBI taxonomy database as reported in the Taxaranks website.</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Depending on the composition of the Kraken DB that you used, you may want to remove unwanted taxa. To do that use <code class="docutils literal notranslate"><span class="pre">grep</span></code>, for example to remove Eukaryota and Viruses:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">grep</span> <span class="o">-</span><span class="n">v</span> <span class="s2">&quot;Viruses\|Eukaryota\|Fungi&quot;</span> <span class="n">abundance_table_IDs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">norm</span><span class="o">.</span><span class="n">taxonomy</span><span class="o">.</span><span class="n">final</span> <span class="o">&gt;</span> <span class="n">abundance_table_IDs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">norm</span><span class="o">.</span><span class="n">final</span><span class="o">.</span><span class="n">Archaea_Bacteria</span>
</pre></div>
</div>
<p>Then, you can generate a new table including only species names (without Eukaryota and Viruses)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">20</span><span class="o">-</span> <span class="n">abundance_table_IDs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">norm</span><span class="o">.</span><span class="n">final</span><span class="o">.</span><span class="n">Archaea_Bacteria</span> <span class="o">&gt;</span> <span class="n">abundance_table_IDs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">norm</span><span class="o">.</span><span class="n">final</span><span class="o">.</span><span class="n">Archaea_Bacteria</span><span class="o">.</span><span class="n">species</span>
</pre></div>
</div>
</div>
</section>
</section>
<section id="sourcetracker">
<h2><span class="section-number">10.2. </span>Sourcetracker<a class="headerlink" href="#sourcetracker" title="Permalink to this headline"></a></h2>
<p><a class="reference external" href="https://github.com/danknights/sourcetracker">Sourcetracker</a> is a tool used to predict the source of microbial communities in a set of input samples (i.e., the sink samples). More info in the <a class="reference external" href="https://www.nature.com/articles/nmeth.1650">publication</a>.</p>
<blockquote>
<div></div></blockquote>
<p>First you must prepare the sourcetracker table, in which the NCBI taxonomic IDs and the abundance values (as integers) are used.
You can parse the table from the genome length normalized abudance table with the R script <code class="docutils literal notranslate"><span class="pre">makeIntegers.R</span></code>, which will generate a <code class="docutils literal notranslate"><span class="pre">filename.int</span></code></p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Rscript</span> <span class="n">makeIntegers</span><span class="o">.</span><span class="n">R</span> <span class="n">abundance_table_IDs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">norm</span>
</pre></div>
</div>
<p>Then you can add the header typical of the sourcetracker format by tweaking the file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sed</span> <span class="s1">&#39;1s/^/#Constructed manually</span><span class="se">\n</span><span class="s1">#OTU id</span><span class="se">\t</span><span class="s1">/&#39;</span> <span class="n">abundance_table_IDs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">norm</span><span class="o">.</span><span class="n">int</span> <span class="o">&gt;</span> <span class="n">abundance_table_IDs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">norm</span><span class="o">.</span><span class="n">int</span><span class="o">.</span><span class="n">sourcetracker</span>
</pre></div>
</div>
<p>After preparing the abundance table, prepare the <strong>mapping file</strong> according to the required <a class="reference external" href="https://github.com/danknights/sourcetracker/blob/master/data/metadata.txt">layout</a>.
In the <strong>SourceSink</strong> column, mark as <code class="docutils literal notranslate"><span class="pre">sink</span></code> the input samples to be tested, and as <code class="docutils literal notranslate"><span class="pre">source</span></code> the reference microbiome samples (e.g. soil, skin, laboratory controls microbiomes).
Here is a <a class="reference external" href="https://docs.google.com/spreadsheets/d/18DmeiKz0nlG3TiqXciVZLiPUREh_BhlE/edit?usp=sharing&amp;ouid=116430744637066257064&amp;rtpof=true&amp;sd=true">dataset</a> of microbiomes with metadata and a template mapping file to be used for the analysis in our lab.</p>
<blockquote>
<div></div></blockquote>
<p>Run sourcetracker using the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Rscript</span> <span class="n">sourcetracker_for_qiime</span><span class="o">.</span><span class="n">r</span> <span class="o">-</span><span class="n">i</span> <span class="n">abundance_table_IDs</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">norm</span><span class="o">.</span><span class="n">int</span><span class="o">.</span><span class="n">sourcetracker</span> <span class="o">-</span><span class="n">m</span> <span class="n">mapping_file</span><span class="o">.</span><span class="n">txt</span> <span class="o">-</span><span class="n">o</span> <span class="n">output_directory</span> <span class="o">-</span><span class="n">v</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The names of the samples in the mapping file (first column) must match those in the abundance table!</p>
</div>
</section>
<section id="analyses-with-phyloseq">
<h2><span class="section-number">10.3. </span>Analyses with Phyloseq<a class="headerlink" href="#analyses-with-phyloseq" title="Permalink to this headline"></a></h2>
<p>We will use the R package <code class="docutils literal notranslate"><span class="pre">Phyloseq</span></code> to explore the microbiome data in the normalized abundance table. Once imported in R, the table can be tweaked to focus on a subset of samples.
Install Phyloseq from <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/phyloseq.html">Bioconductor</a>.</p>
<blockquote>
<div></div></blockquote>
<section id="importing-abundance-tables-in-phyloseq">
<h3><span class="section-number">10.3.1. </span>Importing abundance tables in Phyloseq<a class="headerlink" href="#importing-abundance-tables-in-phyloseq" title="Permalink to this headline"></a></h3>
<p>Activate the following libraries:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">phyloseq</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ggplot2</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">gridExtra</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">vegan</span><span class="p">)</span>
</pre></div>
</div>
<p>Phyloseq works with <code class="docutils literal notranslate"><span class="pre">biom</span></code> objects that store all the required information: abundance of taxa, sample metadata, taxonomy
First you can import the otu table with taxa abundances, including the species names as row names and removing the unwanted columns (the first four resulting from the genome length normalization).</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">otu</span> <span class="o">=</span> <span class="nf">as.matrix</span><span class="p">(</span><span class="nf">read.delim</span><span class="p">(</span><span class="s">&quot;abundance_table_IDs.merged.norm&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span> <span class="n">row.names</span><span class="o">=</span><span class="m">1</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">))</span>
<span class="n">otu.final</span> <span class="o">=</span> <span class="p">(</span><span class="n">otu</span><span class="p">[,</span><span class="o">-</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">4</span><span class="p">)])</span>
<span class="c1"># make the matrix values as numeric</span>
<span class="nf">class</span><span class="p">(</span><span class="n">otu.final</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="s">&quot;numeric&quot;</span>
</pre></div>
</div>
<p>Then you can import the taxonomy file generated by Taxaranks and the <code class="docutils literal notranslate"><span class="pre">getFullTaxaranks.sh</span></code> script (above) and tweak it to remove unwanted columns (1=user_taxa, and 19=species, which is already reported in column 2)</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">taxonomy</span> <span class="o">=</span> <span class="nf">as.matrix</span><span class="p">(</span><span class="nf">read.delim</span><span class="p">(</span><span class="s">&quot;abundance_table_IDs.merged.taxonomy&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span> <span class="n">row.names</span><span class="o">=</span><span class="m">1</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">))</span>
<span class="n">taxonomy.final</span> <span class="o">=</span> <span class="p">(</span><span class="n">taxonomy</span><span class="p">[,</span><span class="o">-</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span><span class="m">19</span><span class="p">)])</span>
</pre></div>
</div>
<p>Then you can generate the biom object with phyloseq.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">phyloseq</span><span class="p">)</span>
<span class="c1"># assing otu and tax tables</span>
<span class="n">OTU</span> <span class="o">=</span> <span class="nf">otu_table</span><span class="p">(</span><span class="n">otu.final</span><span class="p">,</span> <span class="n">taxa_are_rows</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">)</span>
<span class="n">TAX</span> <span class="o">=</span> <span class="nf">tax_table</span><span class="p">(</span><span class="n">taxonomy.final</span><span class="p">)</span>
<span class="n">my_biom</span> <span class="o">=</span> <span class="nf">phyloseq</span><span class="p">(</span><span class="n">OTU</span><span class="p">,</span> <span class="n">TAX</span><span class="p">)</span>
</pre></div>
</div>
<p>As additional step, you can add to the biom file metadata that will be useful to describe the samples and select them in the downstream analyses (e.g. to work on subset of samples).
The metadata are stored in a mapping file, a tab-delimited text file where you have in the first column the names of your samples (matching those in your OTU abundance table), and the other columns as many metadata as you want (e.g. sample source, literature ID etc.).
In the dataset of microbiomes mentioned about you can find metadata to be used for the analysis: <a class="reference external" href="https://docs.google.com/spreadsheets/d/18DmeiKz0nlG3TiqXciVZLiPUREh_BhlE/edit?usp=sharing&amp;ouid=116430744637066257064&amp;rtpof=true&amp;sd=true">Database_metagenomics.xlsx</a></p>
<blockquote>
<div></div></blockquote>
<p>The metadata can be incorporated in your biom file as follows:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># import mapping file with metadata</span>
<span class="n">biom_metadata</span> <span class="o">&lt;-</span> <span class="nf">import_qiime_sample_data</span><span class="p">(</span><span class="s">&quot;map_file_full_dataset_Mar2022.txt&quot;</span><span class="p">)</span>
<span class="c1"># merge data</span>
<span class="n">my_biom</span> <span class="o">&lt;-</span> <span class="nf">merge_phyloseq</span><span class="p">(</span><span class="n">my_biom</span><span class="p">,</span> <span class="n">biom_metadata</span><span class="p">)</span>
</pre></div>
</div>
<p>Finally, here some commands to explore the biom object:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># check sample names:</span>
<span class="nf">sample_data</span><span class="p">(</span><span class="n">my_biom</span><span class="p">)</span>
<span class="c1"># check taxa:</span>
<span class="nf">tax_table</span><span class="p">(</span><span class="n">my_biom</span><span class="p">)</span>
<span class="c1"># check rank names:</span>
<span class="nf">colnames</span><span class="p">(</span><span class="nf">tax_table</span><span class="p">(</span><span class="n">my_biom</span><span class="p">))</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>If needed, it is convenient to rename the taxa columns more appropriately (instead of Rank1 etc.)</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># check the labels used for rank names</span>
<span class="nf">head</span><span class="p">(</span><span class="nf">tax_table</span><span class="p">(</span><span class="n">my_biom</span><span class="p">))</span>
<span class="c1"># change them if needed</span>
<span class="nf">colnames</span><span class="p">(</span><span class="nf">tax_table</span><span class="p">(</span><span class="n">my_biom</span><span class="p">))</span> <span class="o">&lt;-</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;Kingdom&quot;</span><span class="p">,</span> <span class="s">&quot;Phylum&quot;</span><span class="p">,</span> <span class="s">&quot;Class&quot;</span><span class="p">,</span> <span class="s">&quot;Order&quot;</span><span class="p">,</span> <span class="s">&quot;Family&quot;</span><span class="p">,</span> <span class="s">&quot;Genus&quot;</span><span class="p">,</span> <span class="s">&quot;Species&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<p>Remove low-abundance taxa below a desired threshold (most commonly 0.02% is chosen).</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">minTotRelAbun</span> <span class="o">=</span> <span class="m">0.0002</span>
<span class="n">x</span> <span class="o">=</span> <span class="nf">taxa_sums</span><span class="p">(</span><span class="n">my_biom</span><span class="p">)</span>
<span class="n">keepTaxa</span> <span class="o">=</span> <span class="nf">taxa_names</span><span class="p">(</span><span class="n">my_biom</span><span class="p">)[</span><span class="nf">which</span><span class="p">((</span><span class="n">x</span> <span class="o">/</span> <span class="nf">sum</span><span class="p">(</span><span class="n">x</span><span class="p">))</span> <span class="o">&gt;</span> <span class="n">minTotRelAbun</span><span class="p">)]</span>
<span class="n">my_biom_flt</span> <span class="o">=</span> <span class="nf">prune_taxa</span><span class="p">(</span><span class="n">keepTaxa</span><span class="p">,</span> <span class="n">my_biom</span><span class="p">)</span>
</pre></div>
</div>
<p>Next, convert the absolute abundance  values in relative abundance.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">my_biom_rel</span> <span class="o">=</span> <span class="nf">transform_sample_counts</span><span class="p">(</span><span class="n">my_biom_flt</span><span class="p">,</span> <span class="nf">function</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="n">x</span> <span class="o">/</span> <span class="nf">sum</span><span class="p">(</span><span class="n">x</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="barplot-of-taxa-abundances">
<h3><span class="section-number">10.3.2. </span>Barplot of taxa abundances<a class="headerlink" href="#barplot-of-taxa-abundances" title="Permalink to this headline"></a></h3>
<p>We can generate barplots to display the relative abundance of each sample at different taxonomic ranks.
To do that we will first merge the abundance of the taxa of our table (e.g. species) to a higher taxonomic rank (e.g. Phylum).</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># merge taxa at the Phylum rank with tax_glom (in Phyloseq)</span>
<span class="n">my_biom_rel_phylum</span> <span class="o">=</span> <span class="nf">tax_glom</span><span class="p">(</span><span class="n">my_biom_rel</span><span class="p">,</span> <span class="s">&quot;Phylum&quot;</span><span class="p">)</span>
<span class="c1"># plot the relative abundances (you can use ggplot2 themes to improve the chart)</span>
<span class="nf">plot_bar</span><span class="p">(</span><span class="n">my_biom_rel_phylum</span><span class="p">,</span> <span class="n">fill</span> <span class="o">=</span> <span class="s">&quot;Phylum&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span><span class="o">=</span><span class="s">&quot;bottom&quot;</span><span class="p">,</span> <span class="n">axis.text.x</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">4</span><span class="p">,</span> <span class="n">angle</span> <span class="o">=</span> <span class="m">90</span><span class="p">,</span> <span class="n">hjust</span> <span class="o">=</span> <span class="m">1</span><span class="p">))</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>In the <code class="docutils literal notranslate"><span class="pre">tax_glom</span></code> command the label of the rank must match the one that is defined in your original biom file, as reported in the column names of the tax_table command (see above).</p>
</div>
<p>Try to generate barplots at different taxonomic ranks (e.g. family).
To save the plot in the pdf format you can use this command:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">dev.print</span><span class="p">(</span><span class="n">pdf</span><span class="p">,</span> <span class="s">&#39;filename.pdf)&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>You can also focus your analysis on a subset of samples and use the metadata contained in your original mapping file. It would be convenient to add metadata the way it more convenient to select and compare your samples. Here we subsample for some groups of individuals as defined in the metadata column <strong>Group2</strong>.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># subsample single group phyla</span>
<span class="n">subsample</span> <span class="o">=</span> <span class="nf">subset_samples</span><span class="p">(</span><span class="n">my_biom_rel_phylum</span><span class="p">,</span> <span class="n">Group2</span><span class="o">==</span><span class="s">&quot;Portugal calculus&quot;</span><span class="p">)</span>
<span class="c1"># subsample multiple groups phyla</span>
<span class="n">subgroup</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;Portugal calculus&quot;</span><span class="p">,</span><span class="s">&quot;Espinoso&quot;</span><span class="p">,</span><span class="s">&quot;NTC&quot;</span><span class="p">)</span>
<span class="n">subsample</span> <span class="o">=</span> <span class="nf">subset_samples</span><span class="p">(</span><span class="n">my_biom_rel_phylum</span><span class="p">,</span> <span class="n">Group2</span> <span class="o">%in%</span> <span class="n">subgroup</span><span class="p">)</span>
<span class="c1"># barplots in groups with facet_grid</span>
<span class="nf">plot_bar</span><span class="p">(</span><span class="n">subsample</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="s">&quot;phylum&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="nf">facet_grid</span><span class="p">(</span><span class="o">~</span><span class="n">Group2</span><span class="p">,</span> <span class="n">scales</span><span class="o">=</span> <span class="s">&quot;free_x&quot;</span><span class="p">,</span> <span class="n">space</span> <span class="o">=</span> <span class="s">&quot;free_x&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="nf">theme</span><span class="p">(</span><span class="n">legend.position</span><span class="o">=</span><span class="s">&quot;bottom&quot;</span><span class="p">,</span> <span class="n">axis.text.x</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">6</span><span class="p">,</span> <span class="n">angle</span> <span class="o">=</span> <span class="m">90</span><span class="p">,</span> <span class="n">hjust</span> <span class="o">=</span> <span class="m">1</span><span class="p">))</span>
</pre></div>
</div>
</section>
<section id="principal-coordinate-analysis-multidimensional-scaling">
<span id="pcoa"></span><h3><span class="section-number">10.3.3. </span>Principal Coordinate Analysis (Multidimensional Scaling)<a class="headerlink" href="#principal-coordinate-analysis-multidimensional-scaling" title="Permalink to this headline"></a></h3>
<p>The Principal Coordinate Analysis (PCoA), also referred to as metric Multidimensional Scaling (MDS) is a multivariate reduction method performed on distance (or dissimilarity) indexes. Here we will use the Bray-Curtis dissimilarity.
Read more about the MDS and so-called <a class="reference external" href="https://ourcodingclub.github.io/tutorials/ordination/">ordination methods</a></p>
<blockquote>
<div></div></blockquote>
<p>To make the MDS you can focus on a subset of samples in your species abundance table and run the following commands:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># subsample multiple groups phyla</span>
<span class="n">subgroup</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;Portugal calculus&quot;</span><span class="p">,</span><span class="s">&quot;Espinoso&quot;</span><span class="p">,</span><span class="s">&quot;Prehistoric human&quot;</span><span class="p">,</span><span class="s">&quot;NTC&quot;</span><span class="p">,</span><span class="s">&quot;Soil&quot;</span><span class="p">,</span><span class="s">&quot;Skin&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">)</span>
<span class="n">subsample</span> <span class="o">=</span> <span class="nf">subset_samples</span><span class="p">(</span><span class="n">my_biom_rel</span><span class="p">,</span> <span class="n">Group2</span> <span class="o">%in%</span> <span class="n">subgroup</span><span class="p">)</span>
<span class="c1"># calculate Bray-Curtis (BC) dissimilarities</span>
<span class="n">distBC</span> <span class="o">=</span> <span class="nf">distance</span><span class="p">(</span><span class="n">subsample</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s">&quot;bray&quot;</span><span class="p">)</span>
<span class="c1"># make PCoA ordination (MDS) with BC dissimilarities</span>
<span class="n">ordBC</span> <span class="o">=</span> <span class="nf">ordinate</span><span class="p">(</span><span class="n">subsample</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s">&quot;PCoA&quot;</span><span class="p">,</span> <span class="n">distance</span> <span class="o">=</span> <span class="n">distBC</span><span class="p">)</span>
<span class="c1"># two-dimension plot PCoA (MDS) setting the colors of the points automatically as defined in the metadata &quot;Group2&quot;.</span>
<span class="nf">plot_ordination</span><span class="p">(</span><span class="n">subsample</span><span class="p">,</span> <span class="n">ordBC</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s">&quot;Group2&quot;</span><span class="p">)</span>

<span class="c1"># you could play with more plotting options and customizations with ggplot2</span>
<span class="nf">plot_ordination</span><span class="p">(</span><span class="n">my_biom_rel</span><span class="p">,</span> <span class="n">ordBC</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s">&quot;Group2&quot;</span><span class="p">)</span> <span class="o">+</span> <span class="nf">geom_point</span><span class="p">(</span><span class="n">mapping</span> <span class="o">=</span> <span class="nf">aes</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="n">xxx</span><span class="p">,</span> <span class="n">shape</span> <span class="o">=</span> <span class="nf">factor</span><span class="p">(</span><span class="n">yyy</span><span class="p">)))</span> <span class="o">+</span> <span class="nf">ggtitle</span><span class="p">(</span><span class="s">&quot;PCoA: Bray-Curtis&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>You can repeat the analysis at a higher taxonomic ranks (e.g. genus) and see the difference in the plot.</p>
</section>
<section id="non-metric-multidimensional-scaling">
<span id="nmds"></span><h3><span class="section-number">10.3.4. </span>Non-Metric Multidimensional Scaling<a class="headerlink" href="#non-metric-multidimensional-scaling" title="Permalink to this headline"></a></h3>
<p>Unlike other ordination techniques that rely on (primarily Euclidean) distances, such as Principal Coordinates Analysis, nMDS uses rank orders (so not the abundances).
The use of ranks omits some of the issues associated with using absolute distance (e.g., sensitivity to transformation), and as a result is much more flexible technique that accepts a variety of types of data. (It’s also where the “non-metric” part of the name comes from.)
To begin, NMDS requires a distance matrix, or a matrix of dissimilarities. Raw Euclidean distances are not ideal for this purpose: they’re sensitive to total abundances. Consequently, ecologists use the Bray-Curtis dissimilarity calculation.
NMDS arranges points to maximize rank-order correlation between real-world distance and ordination space distance.
You can read more on the nMDS in the blogs curated by ecologists: <a class="reference external" href="https://jonlefcheck.net/2012/10/24/nmds-tutorial-in-r/">link1</a> <a class="reference external" href="https://archetypalecology.wordpress.com/2018/02/18/non-metric-multidimensional-scaling-nmds-what-how/">link2</a></p>
<blockquote>
<div></div></blockquote>
<p>The technique uses a <em>trial and error</em> to find the best positioning in dimensional space. Goodness-of-fit is measured by <strong>stress</strong> – a measure of rank-order disagreement between observed and fitted distances.
You can follow the same steps as above, just change the ordinatiom method supported in the command.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># subsample multiple groups phyla</span>
<span class="n">subgroup</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;Portugal calculus&quot;</span><span class="p">,</span><span class="s">&quot;Espinoso&quot;</span><span class="p">,</span><span class="s">&quot;Prehistoric human&quot;</span><span class="p">,</span><span class="s">&quot;NTC&quot;</span><span class="p">,</span><span class="s">&quot;Soil&quot;</span><span class="p">,</span><span class="s">&quot;Skin&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">)</span>
<span class="n">subsample</span> <span class="o">=</span> <span class="nf">subset_samples</span><span class="p">(</span><span class="n">my_biom_rel</span><span class="p">,</span> <span class="n">Group2</span> <span class="o">%in%</span> <span class="n">subgroup</span><span class="p">)</span>
<span class="c1"># calculate Bray-Curtis (BC) dissimilarities</span>
<span class="n">distBC</span> <span class="o">=</span> <span class="nf">distance</span><span class="p">(</span><span class="n">subsample</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s">&quot;bray&quot;</span><span class="p">)</span>
<span class="c1"># make PCoA ordination (MDS) with BC dissimilarities</span>
<span class="n">ordBC</span> <span class="o">=</span> <span class="nf">ordinate</span><span class="p">(</span><span class="n">subsample</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s">&quot;NMDS&quot;</span><span class="p">,</span> <span class="n">distance</span> <span class="o">=</span> <span class="n">distBC</span><span class="p">)</span>
<span class="c1"># two-dimension plot PCoA (MDS) setting the colors of the points automatically as defined in the metadata &quot;Group2&quot;.</span>
<span class="nf">plot_ordination</span><span class="p">(</span><span class="n">subsample</span><span class="p">,</span> <span class="n">ordBC</span><span class="p">,</span> <span class="n">color</span> <span class="o">=</span> <span class="s">&quot;Group2&quot;</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="differential-taxonomic-abundances-with-deseq2">
<h3><span class="section-number">10.3.5. </span>Differential taxonomic abundances with DESeq2<a class="headerlink" href="#differential-taxonomic-abundances-with-deseq2" title="Permalink to this headline"></a></h3>
<p>A common goal of microbiome studies is to identify differentially abundant taxa (species, genera etc.) between different groups of samples.
One of the tools used to do that is DESeq2, which was originally developed to identify differentially expressed genes in RNAseq data, but it commonly adopted also in microbiome studies. You can read more about DESeq2 in the <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8">original publication</a> and in the <a class="reference external" href="https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html">dedicated page</a> of Bioconductor.
You can install <a class="reference external" href="https://bioconductor.org/packages/release/bioc/html/phyloseq.html">DESeq2 from Bioconductor</a>.</p>
<blockquote>
<div></div></blockquote>
<p>Run DESeq2 on the raw abundace data (<a class="reference external" href="https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#why-un-normalized-counts">here is why</a>), filtered for the low-abundance taxa . You can choose to work on a subset of sample, as done above.</p>
<blockquote>
<div></div></blockquote>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">DESeq2</span><span class="p">)</span>
<span class="c1"># subsample multiple groups phyla</span>
<span class="n">subgroup</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;Portugal calculus&quot;</span><span class="p">,</span><span class="s">&quot;Prehistoric human&quot;</span><span class="p">,</span><span class="s">&quot;NTC&quot;</span><span class="p">,</span><span class="s">&quot;Soil&quot;</span><span class="p">,</span><span class="s">&quot;Skin&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">)</span>
<span class="n">subsample</span> <span class="o">=</span> <span class="nf">subset_samples</span><span class="p">(</span><span class="n">my_biom_flt</span><span class="p">,</span> <span class="n">Group2</span> <span class="o">%in%</span> <span class="n">subgroup</span><span class="p">)</span>
</pre></div>
</div>
<p>DESeq2 does not tolerate zero-counts, for this reason an offset (+1) is commonly applied to remove all zeroes from your table.
After that, run DESeq2 on your table and contrat two sets of samples to find differential taxa abundances.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># make a +1 offset to run Deseq2 (which does not tolerate zero counts)</span>
<span class="nf">otu_table</span><span class="p">(</span><span class="n">subsample</span><span class="p">)</span> <span class="o">=</span> <span class="nf">otu_table</span><span class="p">(</span><span class="n">subsample</span><span class="p">)</span><span class="m">+1</span>
<span class="c1"># make deseq object from phyloseq object</span>
<span class="n">ds</span> <span class="o">=</span> <span class="nf">phyloseq_to_deseq2</span><span class="p">(</span><span class="n">subsample</span><span class="p">,</span> <span class="o">~</span> <span class="n">Group2</span><span class="p">)</span>
<span class="c1"># Run DESeq2</span>
<span class="n">dds.data</span> <span class="o">=</span> <span class="nf">DESeq</span><span class="p">(</span><span class="n">ds</span><span class="p">)</span>
<span class="c1"># With the &#39;contrast&#39; function you screen two different set of samples (based on your metadata) for differential taxa.</span>
<span class="n">res</span> <span class="o">=</span> <span class="nf">results</span><span class="p">(</span><span class="n">dds.data</span><span class="p">,</span> <span class="n">contrast</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;Group2&quot;</span><span class="p">,</span><span class="s">&quot;Portugal calculus&quot;</span><span class="p">,</span><span class="s">&quot;NTC&quot;</span><span class="p">))</span>
</pre></div>
</div>
<p>Then you can explore your results, filter and sort the differential taxa detected based on a False Dicovery Rate (FDR) threshold (e.g. set to 0.01), the log2-fold-change and the base mean. Read more about the <a class="reference external" href="https://www.biostars.org/p/209118/">FDR threshold here</a>.</p>
<blockquote>
<div></div></blockquote>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># sort based on p-value adjusted:</span>
<span class="n">resOrdered</span> <span class="o">=</span> <span class="n">res</span><span class="p">[</span><span class="nf">order</span><span class="p">(</span><span class="n">res</span><span class="o">$</span><span class="n">padj</span><span class="p">,</span> <span class="n">na.last</span><span class="o">=</span><span class="kc">NA</span><span class="p">),</span> <span class="p">]</span>
<span class="c1"># set a threshold value for the False Discovery Rate (FDR):</span>
<span class="n">alpha</span> <span class="o">=</span> <span class="m">0.01</span>
<span class="c1"># get only significant taxa based on p-value adjusted (the FDR):</span>
<span class="n">resSig</span> <span class="o">&lt;-</span> <span class="nf">subset</span><span class="p">(</span><span class="n">resOrdered</span><span class="p">,</span> <span class="n">padj</span> <span class="o">&lt;</span> <span class="n">alpha</span><span class="p">)</span>
<span class="c1"># sort significant values based on the log-fold-change:</span>
<span class="n">resfc</span> <span class="o">=</span> <span class="n">resSig</span><span class="p">[</span><span class="nf">order</span><span class="p">(</span><span class="n">resSig</span><span class="o">$</span><span class="n">log2FoldChange</span><span class="p">),]</span>
<span class="c1"># sort significant values based on abundance:</span>
<span class="n">resbm</span> <span class="o">=</span> <span class="n">resSig</span><span class="p">[</span><span class="nf">order</span><span class="p">(</span><span class="n">resSig</span><span class="o">$</span><span class="n">baseMean</span><span class="p">),]</span>
<span class="c1"># save the the tables</span>
<span class="nf">write.table</span><span class="p">(</span><span class="nf">as.data.frame</span><span class="p">(</span><span class="n">resbm</span><span class="p">),</span> <span class="n">file</span><span class="o">=</span><span class="s">&quot;/path/table.tsv&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">,</span> <span class="n">row.names</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span> <span class="n">col.names</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span> <span class="n">quote</span><span class="o">=</span><span class="bp">F</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>A positive log2-fold-change for a comparison of A vs B means that the OTU in A is more abundant than in B (and viceversa).
For example, a log2-fold-change of −1 means that in A the OTU is of 2^−1 = 0.5 less abundant than in B.</p>
</div>
<p>You can plot the log2-fold-changes to visualize for example the number of differential species belonging to different genera (or phyla) in the two groups contrasted, highlitingh the phylum.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># generate a data frame to handle the data</span>
<span class="n">resSigMod</span> <span class="o">=</span> <span class="nf">cbind</span><span class="p">(</span><span class="nf">as</span><span class="p">(</span><span class="n">resSig</span><span class="p">,</span> <span class="s">&quot;data.frame&quot;</span><span class="p">),</span> <span class="nf">as</span><span class="p">(</span><span class="nf">tax_table</span><span class="p">(</span><span class="n">subsample</span><span class="p">)[</span><span class="nf">rownames</span><span class="p">(</span><span class="n">resSig</span><span class="p">),</span> <span class="p">],</span> <span class="s">&quot;matrix&quot;</span><span class="p">))</span>
<span class="c1"># Plot log-fold-changes of otus based on Genus</span>
<span class="nf">ggplot</span><span class="p">(</span><span class="n">resSigMod</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">genus</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">log2FoldChange</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">phylum</span><span class="p">))</span> <span class="o">+</span>
    <span class="nf">geom_jitter</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="m">3</span><span class="p">,</span> <span class="n">width</span> <span class="o">=</span> <span class="m">0.2</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">theme</span><span class="p">(</span><span class="n">axis.text.x</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">5</span><span class="p">,</span> <span class="n">angle</span> <span class="o">=</span> <span class="m">-90</span><span class="p">,</span> <span class="n">hjust</span> <span class="o">=</span> <span class="m">0</span><span class="p">,</span> <span class="n">vjust</span><span class="o">=</span><span class="m">0.5</span><span class="p">))</span>
<span class="c1"># Plot log-fold-changes of tus based on Phylum</span>
<span class="nf">ggplot</span><span class="p">(</span><span class="n">resSigMod</span><span class="p">,</span> <span class="nf">aes</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="n">phylum</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="n">log2FoldChange</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="n">phylum</span><span class="p">))</span> <span class="o">+</span>
    <span class="nf">geom_jitter</span><span class="p">(</span><span class="n">size</span><span class="o">=</span><span class="m">3</span><span class="p">,</span> <span class="n">width</span> <span class="o">=</span> <span class="m">0.2</span><span class="p">)</span> <span class="o">+</span>
    <span class="nf">theme</span><span class="p">(</span><span class="n">axis.text.x</span> <span class="o">=</span> <span class="nf">element_text</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">10</span><span class="p">,</span> <span class="n">angle</span> <span class="o">=</span> <span class="m">-90</span><span class="p">,</span> <span class="n">hjust</span> <span class="o">=</span> <span class="m">0</span><span class="p">,</span> <span class="n">vjust</span><span class="o">=</span><span class="m">0.5</span><span class="p">))</span>
</pre></div>
</div>
<p>You can also generate a heatmap of the transformed abundance data. For visualization or clustering purposes it might be useful to work with transformed versions of the count data.
Various <a class="reference external" href="https://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#data-transformations-and-visualization">transformations</a> may be applied on the data:
1) ntd: Log transformation of the the data
2) vsd: Variance Stabilizing Transformation
3) rld: Regularized log transformation</p>
<blockquote>
<div></div></blockquote>
<p>When dealing with many samples it is useful to use the vst. To make the various transformatinos run the folowing commands:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># variance stabilizing transformation</span>
<span class="n">vsd</span> <span class="o">=</span> <span class="nf">varianceStabilizingTransformation</span><span class="p">(</span><span class="n">dds.data</span><span class="p">,</span> <span class="n">blind</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span>
<span class="c1"># log transformation</span>
<span class="n">ntd</span> <span class="o">&lt;-</span> <span class="nf">normTransform</span><span class="p">(</span><span class="n">dds.data</span><span class="p">)</span>
<span class="c1"># regularized log transformation</span>
<span class="n">rld</span> <span class="o">&lt;-</span> <span class="nf">rlog</span><span class="p">(</span><span class="n">dds.data</span><span class="p">,</span> <span class="n">blind</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">)</span>
</pre></div>
</div>
<p>We will plot the heatmap of the transformed data only for a reduced number of taxa (with decreasing counts).</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="s">&quot;pheatmap&quot;</span><span class="p">)</span>
<span class="c1">#adjust max number of taxa to display.</span>
<span class="n">select</span> <span class="o">&lt;-</span> <span class="nf">order</span><span class="p">(</span><span class="nf">rowMeans</span><span class="p">(</span><span class="nf">counts</span><span class="p">(</span><span class="n">dds.data</span><span class="p">,</span><span class="n">normalized</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)),</span>
                <span class="n">decreasing</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)[</span><span class="m">1</span><span class="o">:</span><span class="m">50</span><span class="p">]</span>
<span class="c1"># create a dataframe importing using the metadata (e.g. Group2) and incorporating the taxa names as row names.</span>
<span class="n">df</span> <span class="o">&lt;-</span> <span class="nf">as.data.frame</span><span class="p">(</span><span class="nf">colData</span><span class="p">(</span><span class="n">dds.data</span><span class="p">)[,</span><span class="s">&quot;Group2&quot;</span><span class="p">])</span>
<span class="nf">rownames</span><span class="p">(</span><span class="n">df</span><span class="p">)</span> <span class="o">&lt;-</span> <span class="nf">colnames</span><span class="p">(</span><span class="n">subsample</span><span class="o">@</span><span class="n">otu_table</span><span class="p">)</span>
<span class="c1"># plot heatmap (on ntd transformation)</span>
<span class="nf">pheatmap</span><span class="p">(</span><span class="nf">assay</span><span class="p">(</span><span class="n">vsd</span><span class="p">)[</span><span class="n">select</span><span class="p">,],</span> <span class="n">cluster_rows</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span> <span class="n">show_rownames</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">,</span>
         <span class="n">cluster_cols</span><span class="o">=</span><span class="kc">FALSE</span><span class="p">,</span> <span class="n">annotation_col</span><span class="o">=</span><span class="n">df</span><span class="p">,</span> <span class="n">fontsize</span><span class="o">=</span><span class="m">4</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="principal-component-analysis-pca">
<h3><span class="section-number">10.3.6. </span>Principal Component Analysis (PCA)<a class="headerlink" href="#principal-component-analysis-pca" title="Permalink to this headline"></a></h3>
<p>The Principal Component Analysis (PCA) is another multivariate reduction method used for data visualization. You can read more about the general principles of PCA here: <a class="reference external" href="https://builtin.com/data-science/step-step-explanation-principal-component-analysis">link3</a>, <a class="reference external" href="https://archetypalecology.wordpress.com/2018/02/17/principal-component-analysis-in-r/">link4</a></p>
<blockquote>
<div></div></blockquote>
<p>PCA is performed on the abundance data, however an important concept to keep in mind is that <strong>microbiome datasets are compositional</strong> because they have an arbitrary total imposed by the sequencing instrument. This is due to the fact that high-throughput sequencing (HTS) instruments can deliver reads only up to the capacity of the instrument.
Thus, the total read count observed in a HTS run is a fixed-size, random sample of the relative abundance of the molecules in the underlying ecosystem (e.g. oral microbiota).
Several methods applied include count-based strategies (normalized to a constant area or volume, e.g. the sequencing lane output) such as Bray-Curtis dissimilarity, zero-inflated Gaussian models and negative binomial models, but these do not account for the limitations imposed by the instrument and the so-called principle of true indpendence of species in ecology.
Read more about compositonal data in the paper by <a class="reference external" href="https://www.frontiersin.org/articles/10.3389/fmicb.2017.02224/full">Gloor et al. (2017)</a>.</p>
<blockquote>
<div></div></blockquote>
<p>Due to the compositional nature of microbiome datasets, the centered log-ratio (clr) transformation introduced by Aitchison (1986) is often used. The clr-transformed values are scale-invariant, which means that the same ratio is expected to be obtained in a sample with few read counts or an identical sample with many read counts, only the precision of the clr estimate is affected.
The clr transformation uses the geometric mean of the sample vector as the reference. In particular, sample vectors undergo a transformation based on the logarithm of the ratio between the individual elements and the geometric mean of the vector:</p>
<div class="math notranslate nohighlight">
\[g(x) = \sqrt[D]{x_{i}, ...,  x_{D}}\]</div>
<p>Log-ratio transformation is applied to each subject vector <em>x</em> with <em>D</em> features (OTUs). Basically, the analysis of clr data will reveal how OTUs behave relative to the per-sample average.</p>
<div class="math notranslate nohighlight">
\[clr(x) = [\ln{\frac {x_{i}}{g(x)}},...,\ln{\frac {x_{D}}{g(x)}}]\]</div>
<p>We will accomplish the PCA of oral microbiota only (while for comparing different microbiota we used the <a class="reference internal" href="#pcoa"><span class="std std-ref">PCoA</span></a>, or the <a class="reference internal" href="#nmds"><span class="std std-ref">nMDS</span></a>).
For this reason we will subsample the original biom abundance table exclusively for oral groups, for example:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">subgroup</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;Human plaque&quot;</span><span class="p">,</span><span class="s">&quot;Chimpanzee&quot;</span><span class="p">,</span><span class="s">&quot;Prehistoric human&quot;</span><span class="p">,</span><span class="s">&quot;Historic human&quot;</span><span class="p">,</span><span class="s">&quot;Modern human&quot;</span><span class="p">)</span>
<span class="n">subsample</span> <span class="o">=</span> <span class="nf">subset_samples</span><span class="p">(</span><span class="n">my_biom</span><span class="p">,</span> <span class="n">Group2</span> <span class="o">%in%</span> <span class="n">subgroup</span><span class="p">)</span>
</pre></div>
</div>
<p>Then, we filter out low-abundance taxa with a mean &lt; 0.02%.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">minTotRelAbun</span> <span class="o">=</span> <span class="m">0.0002</span>
<span class="n">x</span> <span class="o">=</span> <span class="nf">taxa_sums</span><span class="p">(</span><span class="n">subsample</span><span class="p">)</span>
<span class="n">keepTaxa</span> <span class="o">=</span> <span class="nf">taxa_names</span><span class="p">(</span><span class="n">subsample</span><span class="p">)[</span><span class="nf">which</span><span class="p">((</span><span class="n">x</span> <span class="o">/</span> <span class="nf">sum</span><span class="p">(</span><span class="n">x</span><span class="p">))</span> <span class="o">&gt;</span> <span class="n">minTotRelAbun</span><span class="p">)]</span>
<span class="n">subsample_flt</span> <span class="o">=</span> <span class="nf">prune_taxa</span><span class="p">(</span><span class="n">keepTaxa</span><span class="p">,</span> <span class="n">subsample</span><span class="p">)</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The removal of low-abundance taxa is performed only <strong>after</strong> subsampling the abudance table. Doing this before will keep unwanted taxa that are more common in other microbiota (e.g. soil) and irrelevant for describing oral envrionments (or the specific environment you are analysing with the PCA).</p>
</div>
<p>We will do the clr-transformation of the otu table in the biom object with the package <cite>microbiome</cite>. To install the package:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">BiocManager</span><span class="p">)</span>
<span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&quot;microbiome&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Then activate the library and transform the data with the following command:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">library</span><span class="p">(</span><span class="n">microbiome</span><span class="p">)</span>
<span class="n">subsample_flt_clr</span> <span class="o">=</span> <span class="nf">transform</span><span class="p">(</span><span class="n">subsample_flt</span><span class="p">,</span><span class="s">&quot;clr&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>To refine the analysis you can remove unwanted samples (e.g. those that appeared to be contaminated from Sourcetracker).</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">remove</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;VK1&quot;</span><span class="p">,</span><span class="s">&quot;CS45&quot;</span><span class="p">,</span><span class="s">&quot;CS47&quot;</span><span class="p">,</span><span class="s">&quot;CS48&quot;</span><span class="p">,</span><span class="s">&quot;Jomon_5&quot;</span><span class="p">)</span>
<span class="n">subsample_pruned</span> <span class="o">=</span> <span class="nf">prune_samples</span><span class="p">(</span><span class="o">!</span><span class="p">(</span><span class="n">subsample_flt_clr</span><span class="o">@</span><span class="n">sam_data</span><span class="o">$</span><span class="n">Sample_short</span> <span class="o">%in%</span> <span class="n">remove</span><span class="p">),</span> <span class="n">subsample_flt_clr</span><span class="p">)</span>
</pre></div>
</div>
<p>Finally, make the PCA with the command <cite>ordinate</cite>, adopting the ordination method <cite>RDA</cite> (redundancy analysis), which without constraints correspond to the PCA in phyloseq.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">ord_clr</span> <span class="o">=</span> <span class="nf">ordinate</span><span class="p">(</span><span class="n">subsample_pruned</span><span class="p">,</span> <span class="s">&quot;RDA&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>You can plot the variance associated with each PC:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">plot_scree</span><span class="p">(</span><span class="n">ord_clr</span><span class="p">)</span> <span class="o">+</span>
<span class="nf">geom_bar</span><span class="p">(</span><span class="n">stat</span><span class="o">=</span><span class="s">&quot;identity&quot;</span><span class="p">,</span> <span class="n">fill</span> <span class="o">=</span> <span class="s">&quot;blue&quot;</span><span class="p">)</span> <span class="o">+</span>
<span class="nf">labs</span><span class="p">(</span><span class="n">x</span> <span class="o">=</span> <span class="s">&quot;\nAxis&quot;</span><span class="p">,</span> <span class="n">y</span> <span class="o">=</span> <span class="s">&quot;Proportion of Variance\n&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>And finally plot the PCA in a 2D chart.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">plot_ordination</span><span class="p">(</span><span class="n">subsample_flt_clr</span><span class="p">,</span> <span class="n">ord_clr</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s">&quot;Group2&quot;</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">geom_point</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">2</span><span class="p">)</span>
</pre></div>
</div>
<p>To display the labels of each symbol:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">plot_ordination</span><span class="p">(</span><span class="n">subsample_flt_clr</span><span class="p">,</span> <span class="n">ord_clr</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s">&quot;Group2&quot;</span><span class="p">,</span> <span class="n">label</span><span class="o">=</span><span class="s">&quot;Sample_short&quot;</span><span class="p">)</span> <span class="o">+</span>
  <span class="nf">geom_point</span><span class="p">(</span><span class="n">size</span> <span class="o">=</span> <span class="m">2</span><span class="p">)</span>
</pre></div>
</div>
<p><strong>ALTERNATIVE:</strong> Another package used to perform the PCA is <cite>mixOmics</cite>. To install it:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1">## install BiocManager if not installed</span>
<span class="nf">if </span><span class="p">(</span><span class="o">!</span><span class="nf">require</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">,</span> <span class="n">quietly</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">))</span>
    <span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;BiocManager&quot;</span><span class="p">)</span>
<span class="c1">## install mixOmics</span>
<span class="n">BiocManager</span><span class="o">::</span><span class="nf">install</span><span class="p">(</span><span class="s">&#39;mixOmics&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>Here are the commands to make the PCA with mixOmics. Note the conversion to count-per-million and the +1 offest that is necessary to remove zero-counts (which cannot be handled by the clr-transformation of mixOmics).</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="c1"># remove contaminated samples</span>
<span class="n">remove</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;VK1&quot;</span><span class="p">,</span><span class="s">&quot;CS45&quot;</span><span class="p">,</span><span class="s">&quot;CS47&quot;</span><span class="p">,</span><span class="s">&quot;CS48&quot;</span><span class="p">,</span><span class="s">&quot;Jomon_5&quot;</span><span class="p">)</span>
<span class="n">subsample_pruned</span> <span class="o">=</span> <span class="nf">prune_samples</span><span class="p">(</span><span class="o">!</span><span class="p">(</span><span class="n">subsample_flt</span><span class="o">@</span><span class="n">sam_data</span><span class="o">$</span><span class="n">Sample_short</span> <span class="o">%in%</span> <span class="n">remove</span><span class="p">),</span> <span class="n">subsample_flt</span><span class="p">)</span>
<span class="c1"># isolate the otu_table of the biom as separate object.</span>
<span class="n">tab</span> <span class="o">=</span> <span class="nf">otu_table</span><span class="p">(</span><span class="n">subsample_pruned</span><span class="p">)</span>
<span class="c1"># make on offset (+1) of your count data, after converting them to counts-per-million (to make the offset irrelevant)</span>
<span class="n">tab_cpm_off</span> <span class="o">=</span> <span class="p">(</span><span class="n">tab</span><span class="o">*</span><span class="m">1000000</span><span class="p">)</span><span class="m">+1</span>
<span class="n">tab_cpm_off</span> <span class="o">=</span> <span class="nf">t</span><span class="p">(</span><span class="n">tab_cpm_off</span><span class="p">)</span>
<span class="c1"># create e vector from the metadata that you want to use:</span>
<span class="n">group</span> <span class="o">=</span> <span class="n">subsample_pruned</span><span class="o">@</span><span class="n">sam_data</span><span class="o">$</span><span class="n">Group2</span>
<span class="c1"># make the PCA with mixOmics. Note the CLR tranformation called in the command.</span>
<span class="nf">library</span><span class="p">(</span><span class="n">mixOmics</span><span class="p">)</span>
<span class="nf">tune.pca</span><span class="p">(</span><span class="n">tab_cpm_off</span><span class="p">,</span> <span class="n">ncomp</span> <span class="o">=</span> <span class="m">10</span><span class="p">,</span> <span class="n">center</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">scale</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> <span class="n">logratio</span> <span class="o">=</span> <span class="s">&#39;CLR&#39;</span><span class="p">)</span>
<span class="n">pca.res</span> <span class="o">&lt;-</span> <span class="nf">pca</span><span class="p">(</span><span class="n">tab_cpm_off</span><span class="p">,</span> <span class="n">ncomp</span> <span class="o">=</span> <span class="m">10</span><span class="p">,</span> <span class="n">center</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">scale</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> <span class="n">logratio</span> <span class="o">=</span> <span class="s">&#39;CLR&#39;</span><span class="p">)</span>
<span class="c1"># PLot PCA</span>
<span class="nf">par</span><span class="p">(</span><span class="n">mfrow</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span> <span class="m">1</span><span class="p">))</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">pca.res</span><span class="o">$</span><span class="n">variates</span><span class="o">$</span><span class="n">X</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span> <span class="n">pca.res</span><span class="o">$</span><span class="n">variates</span><span class="o">$</span><span class="n">X</span><span class="p">[,</span><span class="m">2</span><span class="p">],</span> <span class="n">type</span><span class="o">=</span><span class="s">&quot;n&quot;</span><span class="p">,</span> <span class="n">main</span><span class="o">=</span><span class="s">&quot;PCA&quot;</span><span class="p">,</span> <span class="n">cex.axis</span><span class="o">=</span><span class="m">0.75</span><span class="p">,</span> <span class="n">cex.lab</span><span class="o">=</span><span class="m">0.75</span><span class="p">,</span> <span class="n">xlab</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">,</span> <span class="n">ylab</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">)</span>
<span class="nf">abline</span><span class="p">(</span><span class="n">v</span><span class="o">=</span><span class="m">0</span><span class="p">,</span> <span class="n">lty</span><span class="o">=</span><span class="m">3</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s">&quot;grey&quot;</span><span class="p">)</span>
<span class="nf">abline</span><span class="p">(</span><span class="n">h</span><span class="o">=</span><span class="m">0</span><span class="p">,</span> <span class="n">lty</span><span class="o">=</span><span class="m">3</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s">&quot;grey&quot;</span><span class="p">)</span>
<span class="c1"># add the points and the symbols (pch) based on the number of groups present in your metadata (those selected when subsampling the oral microbiota)</span>
<span class="nf">points</span><span class="p">(</span><span class="n">pca.res</span><span class="o">$</span><span class="n">variates</span><span class="o">$</span><span class="n">X</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span> <span class="n">pca.res</span><span class="o">$</span><span class="n">variates</span><span class="o">$</span><span class="n">X</span><span class="p">[,</span><span class="m">2</span><span class="p">],</span> <span class="n">cex</span><span class="o">=</span><span class="m">1.2</span><span class="p">,</span> <span class="n">bg</span><span class="o">=</span><span class="nf">factor</span><span class="p">(</span><span class="n">group</span><span class="p">),</span> <span class="n">col</span><span class="o">=</span><span class="nf">factor</span><span class="p">(</span><span class="n">group</span><span class="p">),</span> <span class="n">pch</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">21</span><span class="p">,</span><span class="m">22</span><span class="p">,</span><span class="m">23</span><span class="p">,</span><span class="m">3</span><span class="p">,</span><span class="m">4</span><span class="p">)[</span><span class="nf">as.factor</span><span class="p">(</span><span class="n">group</span><span class="p">)],</span> <span class="n">lwd</span><span class="o">=</span><span class="m">0.7</span><span class="p">)</span>
<span class="nf">legend</span><span class="p">(</span><span class="s">&quot;topleft&quot;</span><span class="p">,</span> <span class="n">legend</span> <span class="o">=</span> <span class="nf">sort</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="n">group</span><span class="p">)),</span> <span class="n">bty</span> <span class="o">=</span> <span class="s">&quot;n&quot;</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="nf">sort</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="nf">as.factor</span><span class="p">(</span><span class="n">group</span><span class="p">))),</span> <span class="n">pt.cex</span><span class="o">=</span><span class="m">1</span><span class="p">,</span> <span class="n">cex</span><span class="o">=</span><span class="m">0.5</span><span class="p">,</span> <span class="n">pt.bg</span><span class="o">=</span><span class="nf">sort</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="nf">as.factor</span><span class="p">(</span><span class="n">group</span><span class="p">))),</span> <span class="n">pch</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">21</span><span class="p">,</span><span class="m">22</span><span class="p">,</span><span class="m">23</span><span class="p">,</span><span class="m">3</span><span class="p">,</span><span class="m">4</span><span class="p">),</span> <span class="n">pt.lwd</span><span class="o">=</span><span class="m">0.6</span><span class="p">)</span>
<span class="nf">text</span><span class="p">(</span><span class="n">pca.res</span><span class="o">$</span><span class="n">variates</span><span class="o">$</span><span class="n">X</span><span class="p">,</span> <span class="n">labels</span> <span class="o">=</span> <span class="n">subsample_pruned</span><span class="o">@</span><span class="n">sam_data</span><span class="o">$</span><span class="n">Sample_short</span><span class="p">,</span> <span class="n">cex</span><span class="o">=</span><span class="m">0.4</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="7_Filtering_SNPs.html" class="btn btn-neutral float-left" title="9. Filtering, annotating and combining SNPs" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Claudio Ottoni.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
    <!-- Theme Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-XXXXXXX-1"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'UA-XXXXXXX-1', {
          'anonymize_ip': false,
      });
    </script> 

</body>
</html>