<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Paleogenomics at Tor Vergata &mdash; Paleogenomics at Tor Vergata 0.1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="1. List of Tools" href="1_ListTools.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="#" class="icon icon-home"> Paleogenomics at Tor Vergata
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="1_ListTools.html">1. List of Tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="2_Terminal_setup.html">2. First step to set-up your machine</a></li>
<li class="toctree-l1"><a class="reference internal" href="2_Useful_commands.html">3. Useful commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="3_Slurm_HPC.html">4. Working on HPC with Slurm</a></li>
<li class="toctree-l1"><a class="reference internal" href="3_ReadsFiltering.html">5. Quality filtering of reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="4_Metagenomics_v2.html">6. Metagenomic screening of shotgun data</a></li>
<li class="toctree-l1"><a class="reference internal" href="5_ReadsMapping_v2.html">7. Alignment of reads to a reference genome</a></li>
<li class="toctree-l1"><a class="reference internal" href="6_VariantsCall.html">8. Variant calling and visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="7_Filtering_SNPs.html">9. Filtering, annotating and combining SNPs</a></li>
<li class="toctree-l1"><a class="reference internal" href="9_Metagenomic_analysis.html">10. Analysis of Metagenomic data</a></li>
<li class="toctree-l1"><a class="reference internal" href="10_AMR_analysis.html">11. Antimicrobial Resistance (AMR) analysis</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">Paleogenomics at Tor Vergata</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="#" class="icon icon-home"></a> &raquo;</li>
      <li>Paleogenomics at Tor Vergata</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <a class="reference internal image-reference" href="_images/Tor-Vergata_logo_EN_cmyk.png"><img alt="_images/Tor-Vergata_logo_EN_cmyk.png" src="_images/Tor-Vergata_logo_EN_cmyk.png" style="width: 400px;" /></a>
<section id="paleogenomics-at-tor-vergata">
<h1>Paleogenomics at Tor Vergata<a class="headerlink" href="#paleogenomics-at-tor-vergata" title="Permalink to this headline"></a></h1>
<p>Tutorial of bioinformatics at the Laboratory of Molecular Anthropology, University of Rome Tor Vergata.</p>
<section id="contents">
<h2>Contents<a class="headerlink" href="#contents" title="Permalink to this headline"></a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="1_ListTools.html">1. List of Tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="2_Terminal_setup.html">2. First step to set-up your machine</a><ul>
<li class="toctree-l2"><a class="reference internal" href="2_Terminal_setup.html#set-up-your-terminal-machine">2.1. Set-up your terminal machine</a></li>
<li class="toctree-l2"><a class="reference internal" href="2_Terminal_setup.html#install-miniconda-to-work-on-your-local-machine">2.2. Install Miniconda to work on your local machine</a></li>
<li class="toctree-l2"><a class="reference internal" href="2_Terminal_setup.html#create-an-environment-with-conda">2.3. Create an environment with Conda</a></li>
<li class="toctree-l2"><a class="reference internal" href="2_Terminal_setup.html#install-packages-inside-the-environment">2.4. Install packages inside the environment</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="2_Useful_commands.html">3. Useful commands</a><ul>
<li class="toctree-l2"><a class="reference internal" href="2_Useful_commands.html#download-fastq-sequencing-data-from-the-sra">3.1. Download FASTQ sequencing data from the SRA</a></li>
<li class="toctree-l2"><a class="reference internal" href="2_Useful_commands.html#download-referece-sequences-from-the-ncbi">3.2. Download referece sequences from the NCBI</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="3_Slurm_HPC.html">4. Working on HPC with Slurm</a><ul>
<li class="toctree-l2"><a class="reference internal" href="3_Slurm_HPC.html#connect-to-the-hpc-cluster">4.1. Connect to the HPC cluster</a></li>
<li class="toctree-l2"><a class="reference internal" href="3_Slurm_HPC.html#run-jobs-with-slurm">4.2. Run jobs with SLURM</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="3_ReadsFiltering.html">5. Quality filtering of reads</a><ul>
<li class="toctree-l2"><a class="reference internal" href="3_ReadsFiltering.html#reads-quality-control">5.1. Reads quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="3_ReadsFiltering.html#reads-quality-filtering">5.2. Reads quality filtering</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="4_Metagenomics_v2.html">6. Metagenomic screening of shotgun data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="4_Metagenomics_v2.html#kraken-2">6.1. Kraken 2</a></li>
<li class="toctree-l2"><a class="reference internal" href="4_Metagenomics_v2.html#bracken">6.2. Bracken</a></li>
<li class="toctree-l2"><a class="reference internal" href="4_Metagenomics_v2.html#mataphlan-3">6.3. Mataphlan 3</a></li>
<li class="toctree-l2"><a class="reference internal" href="4_Metagenomics_v2.html#other-tools">6.4. Other tools</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="5_ReadsMapping_v2.html">7. Alignment of reads to a reference genome</a><ul>
<li class="toctree-l2"><a class="reference internal" href="5_ReadsMapping_v2.html#preparation-of-the-reference-sequence">7.1. Preparation of the reference sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="5_ReadsMapping_v2.html#alignment-of-the-reads-to-the-reference-sequence">7.2. Alignment of the reads to the reference sequence</a></li>
<li class="toctree-l2"><a class="reference internal" href="5_ReadsMapping_v2.html#create-mapping-summary-reports">7.3. Create mapping summary reports</a></li>
<li class="toctree-l2"><a class="reference internal" href="5_ReadsMapping_v2.html#alignment-of-reads-to-circular-reference-genomes">7.4. Alignment of reads to circular reference genomes</a></li>
<li class="toctree-l2"><a class="reference internal" href="5_ReadsMapping_v2.html#damage-analysis-and-quality-rescaling-of-the-bam-file">7.5. Damage analysis and quality rescaling of the BAM file</a></li>
<li class="toctree-l2"><a class="reference internal" href="5_ReadsMapping_v2.html#edit-distance">7.6. Edit Distance</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="6_VariantsCall.html">8. Variant calling and visualization</a><ul>
<li class="toctree-l2"><a class="reference internal" href="6_VariantsCall.html#variants-calling">8.1. Variants calling</a></li>
<li class="toctree-l2"><a class="reference internal" href="6_VariantsCall.html#variants-visualization">8.2. Variants visualization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="7_Filtering_SNPs.html">9. Filtering, annotating and combining SNPs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="7_Filtering_SNPs.html#snps-filtering-and-annotion">9.1. SNPs filtering and annotion</a></li>
<li class="toctree-l2"><a class="reference internal" href="7_Filtering_SNPs.html#compare-and-combine-multiple-annotation-files">9.2. Compare and combine multiple annotation files</a></li>
<li class="toctree-l2"><a class="reference internal" href="7_Filtering_SNPs.html#phylogenetic-tree-reconstruction">9.3. Phylogenetic tree reconstruction</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="9_Metagenomic_analysis.html">10. Analysis of Metagenomic data</a><ul>
<li class="toctree-l2"><a class="reference internal" href="9_Metagenomic_analysis.html#prepare-abundance-tables-for-downstream-analyses">10.1. Prepare abundance tables for downstream analyses</a></li>
<li class="toctree-l2"><a class="reference internal" href="9_Metagenomic_analysis.html#sourcetracker">10.2. Sourcetracker</a></li>
<li class="toctree-l2"><a class="reference internal" href="9_Metagenomic_analysis.html#analyses-with-phyloseq">10.3. Analyses with Phyloseq</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="10_AMR_analysis.html">11. Antimicrobial Resistance (AMR) analysis</a><ul>
<li class="toctree-l2"><a class="reference internal" href="10_AMR_analysis.html#build-the-card-database">11.1. Build the CARD database</a></li>
<li class="toctree-l2"><a class="reference internal" href="10_AMR_analysis.html#build-the-database-of-housekeeping-hk-genes">11.2. Build the database of housekeeping (HK) genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="10_AMR_analysis.html#interrogate-the-card-and-housekeeping-genes-databases">11.3. Interrogate the CARD and Housekeeping genes databases</a></li>
<li class="toctree-l2"><a class="reference internal" href="10_AMR_analysis.html#normalization-of-data">11.4. Normalization of data</a></li>
<li class="toctree-l2"><a class="reference internal" href="10_AMR_analysis.html#amr-data-analysis-in-r">11.5. AMR data analysis in R</a></li>
</ul>
</li>
</ul>
</div>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="1_ListTools.html" class="btn btn-neutral float-right" title="1. List of Tools" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Claudio Ottoni.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
    <!-- Theme Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-XXXXXXX-1"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'UA-XXXXXXX-1', {
          'anonymize_ip': false,
      });
    </script> 

</body>
</html>