<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>10. Analysis of Metagenomic data (R session) &mdash; Paleogenomics at Tor Vergata 0.1.0 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="11. Hands-on" href="8_HandsOn.html" />
    <link rel="prev" title="9. Filtering, annotating and combining SNPs" href="7_Filtering_SNPs.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> Paleogenomics at Tor Vergata
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="1_ListTools.html">1. List of Tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="2_Terminal_setup.html">2. First step to set-up your machine</a></li>
<li class="toctree-l1"><a class="reference internal" href="2_Useful_commands.html">3. Useful commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="3_Slurm_HPC.html">4. Working on HPC with Slurm</a></li>
<li class="toctree-l1"><a class="reference internal" href="3_ReadsFiltering.html">5. Quality filtering of reads</a></li>
<li class="toctree-l1"><a class="reference internal" href="4_Metagenomics_v2.html">6. Metagenomic screening of shotgun data</a></li>
<li class="toctree-l1"><a class="reference internal" href="5_ReadsMapping_v2.html">7. Alignment of reads to a reference genome</a></li>
<li class="toctree-l1"><a class="reference internal" href="6_VariantsCall.html">8. Variant calling and visualization</a></li>
<li class="toctree-l1"><a class="reference internal" href="7_Filtering_SNPs.html">9. Filtering, annotating and combining SNPs</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">10. Analysis of Metagenomic data (R session)</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#abundance-tables-and-genome-length-normalization">10.1. Abundance tables and genome length normalization</a></li>
<li class="toctree-l2"><a class="reference internal" href="#sourcetracker">10.2. Sourcetracker</a></li>
<li class="toctree-l2"><a class="reference internal" href="#analyses-with-phyloseq">10.3. Analyses with Phyloseq</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#r-session">10.3.1. R session</a><ul>
<li class="toctree-l4"><a class="reference internal" href="#import-files-and-prepare-the-abundance-tables">10.3.1.1. Import files and prepare the abundance tables</a></li>
<li class="toctree-l4"><a class="reference internal" href="#upgma">10.3.1.2. UPGMA</a></li>
<li class="toctree-l4"><a class="reference internal" href="#pca">10.3.1.3. PCA</a></li>
</ul>
</li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="8_HandsOn.html">11. Hands-on</a></li>
<li class="toctree-l1"><a class="reference internal" href="9_old_material.html">12. Material from previous courses</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Paleogenomics at Tor Vergata</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li><span class="section-number">10. </span>Analysis of Metagenomic data (R session)</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/9_R_session.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="analysis-of-metagenomic-data-r-session">
<h1><span class="section-number">10. </span>Analysis of Metagenomic data (R session)<a class="headerlink" href="#analysis-of-metagenomic-data-r-session" title="Permalink to this headline"></a></h1>
<p>Here we analyse shotgun sequence data from dental calculus. After filtering and collapsing the reads, classify the reads of each sample with Kraken2 and estimate the abudances with Bracken.
After that, you can use R to run analyses and create charts from the abundance tables generated with Bracken.</p>
<section id="abundance-tables-and-genome-length-normalization">
<h2><span class="section-number">10.1. </span>Abundance tables and genome length normalization<a class="headerlink" href="#abundance-tables-and-genome-length-normalization" title="Permalink to this headline"></a></h2>
<p>Run the <code class="docutils literal notranslate"><span class="pre">brackenToAbundanceTable.R</span></code> R script to merge the species aundances of all the samples in one table. The script needs as argument the path to the folder containing the bracken results.
Move to the bracken results folder and type the following command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Rscript</span> <span class="n">brackenToAbundanceTable</span><span class="o">.</span><span class="n">R</span> <span class="o">.</span>
</pre></div>
</div>
<p>The script will generate two abundance tables, <code class="docutils literal notranslate"><span class="pre">taxa_abundance_bracken_IDs.txt</span></code>, which contains the species names as NCBI IDs, and <code class="docutils literal notranslate"><span class="pre">taxa_abundance_bracken_names.txt</span></code>, which contains the actual species names.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You can generate abundance tables for various datasets (with bracken files contained in dedicated folders) in different folders and rename the final tables based on the folder (i.e. dataset) name. Try the following commands in the the folder containing all the datasets-subfolders:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>for i in $(find -name &quot;*IDs.txt&quot; -type f | sort); do tag=$(basename $(dirname &quot;$i&quot;)); cp $i ${FOLDER}/${tag}_abundance_bracken_IDs.txt; done
for i in $(find -name &quot;*names.txt&quot; -type f | sort); do tag=$(basename $(dirname &quot;$i&quot;)); cp $i ${FOLDER}/${tag}_abundance_bracken_names.txt; done
</pre></div>
</div>
</div>
<p>Then use the <code class="docutils literal notranslate"><span class="pre">abundanceTablesMerger.R</span></code> script to merge all the tables (either _IDs or _names) to obtain one comprehensive table (file.merged) that you will use for downstream analyses:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Rscript</span> <span class="n">abundanceTablesMerger</span><span class="o">.</span><span class="n">R</span> <span class="o">*.</span><span class="n">txt</span>
</pre></div>
</div>
<p>Since the Kraken database was built from complete Bacterial, Archaeal and Viral genomes, we must make a normalization for the genome leghts of each species. This normalization is important when we want to analyse relative taxa abundances to characterise full microbiomes.
To do that will use a Python script that takes three arguments:</p>
<ol class="arabic simple">
<li><p>The abundance table with names</p></li>
<li><p>The table of genome lengths (parsed from the NCBI genome browser <a class="reference external" href="https://www.ncbi.nlm.nih.gov/genome/browse/#!/prokaryotes/">https://www.ncbi.nlm.nih.gov/genome/browse/#!/prokaryotes/</a>)</p></li>
<li><p>The name of the output file</p></li>
</ol>
<p>Run the normalization with <code class="docutils literal notranslate"><span class="pre">gL-normalizer-lite.py</span></code> and type the command:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">python</span> <span class="n">gL</span><span class="o">-</span><span class="n">normalizer</span><span class="o">-</span><span class="n">lite</span><span class="o">.</span><span class="n">py</span> <span class="n">abundance_table_names</span><span class="o">.</span><span class="n">merged</span> <span class="n">prokaryotes_viruses_organelles</span><span class="o">.</span><span class="n">table</span> <span class="n">abundance_table_names</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">norm</span>
</pre></div>
</div>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<p>The script searches the <strong>names</strong> of the species in the file with the list of genome lengths of species in the NCBI and divides the abundance of each species by the length of the assembly available in the NCBI. This means that you can’t run the genome length normalization on the IDs abundance table.</p>
</div>
<p>Open the normalized output table, the layout reports the <strong>Assembly stage</strong>, the <strong>Length in Megabases</strong> and the kind of <strong>Match</strong>, exact, species (when a strain is not found), and genus (when a species is not found).</p>
</section>
<section id="sourcetracker">
<h2><span class="section-number">10.2. </span>Sourcetracker<a class="headerlink" href="#sourcetracker" title="Permalink to this headline"></a></h2>
<p>–&gt; TO TEST
First you must prepare the sourcetracker table, which requires the NCBI taxonomic IDs.
Get the species names from the <cite>abundance_table_names.merged</cite> table in a separate file:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">awk</span> <span class="o">-</span><span class="sa">F</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="s1">&#39;{print $1}&#39;</span> <span class="n">abundance_table_names</span><span class="o">.</span><span class="n">merged</span> <span class="o">|</span> <span class="n">tail</span> <span class="o">-</span><span class="n">n</span> <span class="o">+</span><span class="mi">2</span> <span class="o">&gt;</span> <span class="n">abundance_table_names</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">spList</span>
</pre></div>
</div>
<p>Run <cite>taxaranks</cite> on the species list to obtain the full taxonomic ranks:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">taxaranks</span> <span class="o">-</span><span class="n">i</span> <span class="n">abundance_table_names</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">spList</span> <span class="o">-</span><span class="n">o</span> <span class="n">abundance_table_names</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">spIDs</span> <span class="o">-</span><span class="n">t</span>
</pre></div>
</div>
<p>Isolate the species NCBI IDs (column 36) and save them in a separate file</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">awk</span> <span class="o">-</span><span class="sa">F</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span> <span class="s1">&#39;{print $36}&#39;</span> <span class="n">abundance_table_names</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">spIDs</span> <span class="o">&gt;</span> <span class="n">abundance_table_names</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">IDs</span>
</pre></div>
</div>
<p>Paste the IDs to the original table and remove the species names (column 2)</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">paste</span> <span class="n">abundance_table_names</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">IDs</span> <span class="n">abundance_table_names</span><span class="o">.</span><span class="n">merged</span> <span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f2</span> <span class="o">--</span><span class="n">complement</span> <span class="o">&gt;</span> <span class="n">abundance_table_names</span><span class="o">.</span><span class="n">merged</span><span class="o">.</span><span class="n">sourcetracker</span>
</pre></div>
</div>
<p>Tweak the table to sourcetracker format (add the header). –&gt; MAKE A SCRIPT?</p>
</section>
<section id="analyses-with-phyloseq">
<h2><span class="section-number">10.3. </span>Analyses with Phyloseq<a class="headerlink" href="#analyses-with-phyloseq" title="Permalink to this headline"></a></h2>
<section id="r-session">
<h3><span class="section-number">10.3.1. </span>R session<a class="headerlink" href="#r-session" title="Permalink to this headline"></a></h3>
<section id="import-files-and-prepare-the-abundance-tables">
<h4><span class="section-number">10.3.1.1. </span>Import files and prepare the abundance tables<a class="headerlink" href="#import-files-and-prepare-the-abundance-tables" title="Permalink to this headline"></a></h4>
<p>Download the normalized table <code class="docutils literal notranslate"><span class="pre">taxa_abundance_bracken_names_normalized.txt</span></code> in your local machine (e.g. with <code class="docutils literal notranslate"><span class="pre">scp</span></code>) and open R. Feel free to rename the file at your own convenience (here we will rename it <code class="docutils literal notranslate"><span class="pre">mytable_normalized.txt</span></code>)</p>
<p>First of all, in R, we must set up the folder which contains the abundance table:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">setwd</span><span class="p">(</span><span class="s">&quot;/path/to/your/folder&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Then we can import the abundance table files, setting the 1st column as row names. Then we remove the 1st, 2nd and 3rd column, which will not be used for the analysis.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">mytable</span> <span class="o">=</span> <span class="nf">read.delim</span><span class="p">(</span><span class="s">&quot;mytable_normalized.txt&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span> <span class="n">row.names</span><span class="o">=</span><span class="m">1</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
<span class="n">mytable.final</span> <span class="o">=</span> <span class="n">mytable</span><span class="p">[,</span><span class="o">-</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">3</span><span class="p">)]</span>
</pre></div>
</div>
<p>Then we filter the species in the table for their abundance by removing those that are represented below a threshold of 0.02%.
To do that, we define a function (that we call low.count.removal)</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">low.count.removal</span> <span class="o">=</span> <span class="nf">function</span><span class="p">(</span>
                        <span class="n">data</span><span class="p">,</span>                 <span class="c1"># OTU count data frame of size n (sample) x p (OTU)</span>
                        <span class="n">percent</span><span class="o">=</span><span class="m">0.02</span>  <span class="c1"># cutoff chosen</span>
                        <span class="p">){</span>
    <span class="n">keep.otu</span> <span class="o">=</span> <span class="nf">which</span><span class="p">(</span><span class="nf">colSums</span><span class="p">(</span><span class="n">data</span><span class="p">)</span><span class="o">*</span><span class="m">100</span><span class="o">/</span><span class="p">(</span><span class="nf">sum</span><span class="p">(</span><span class="nf">colSums</span><span class="p">(</span><span class="n">data</span><span class="p">)))</span> <span class="o">&gt;</span> <span class="n">percent</span><span class="p">)</span>
    <span class="n">data.filter</span> <span class="o">=</span> <span class="n">data</span><span class="p">[,</span><span class="n">keep.otu</span><span class="p">]</span>
    <span class="nf">return</span><span class="p">(</span><span class="nf">list</span><span class="p">(</span><span class="n">data.filter</span> <span class="o">=</span> <span class="n">data.filter</span><span class="p">,</span> <span class="n">keep.otu</span> <span class="o">=</span> <span class="n">keep.otu</span><span class="p">))</span>
<span class="p">}</span>
</pre></div>
</div>
<p>We run the function on our table, setting up the threshold at 0.02%. Note that the table is transposed to run the function:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">result.filter</span> <span class="o">=</span> <span class="nf">low.count.removal</span><span class="p">(</span><span class="nf">t</span><span class="p">(</span><span class="n">mytable.final</span><span class="p">),</span> <span class="n">percent</span><span class="o">=</span><span class="m">0.02</span><span class="p">)</span>
</pre></div>
</div>
<p>We generate a table with the filtered data:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">mytable.final.flt</span> <span class="o">=</span> <span class="n">result.filter</span><span class="o">$</span><span class="n">data.filter</span>
</pre></div>
</div>
<p>In the next step, we normalize the reads for the sequencing depth. This means that we will account for the total reads generated for each sample, and normalize the species abundance for that number.
To do that we will define a <strong>Total Sum Squared</strong> function</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">TSS.divide</span> <span class="o">=</span> <span class="nf">function</span><span class="p">(</span><span class="n">x</span><span class="p">){</span>
 <span class="n">x</span><span class="o">/</span><span class="nf">sum</span><span class="p">(</span><span class="n">x</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
<p>The function is applied to the table, and each row must represent a sample. For this reason we transpose the table.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">mytable.final.flt.tss</span> <span class="o">=</span> <span class="nf">t</span><span class="p">(</span><span class="nf">apply</span><span class="p">(</span><span class="n">mytable.final.flt</span><span class="p">,</span> <span class="m">1</span><span class="p">,</span> <span class="n">TSS.divide</span><span class="p">))</span>
</pre></div>
</div>
<p>We have just generated a table of species abundances of ancient dental calculus samples, normalized for genome lenghts and sequencing depths.
For comparative analysis, we can now include in our analysis a dataset of normalized species abundances generated with Minikraken (version <code class="docutils literal notranslate"><span class="pre">minikraken2_v1_8GB_201904</span></code>) representing other microbiomes.
To do that, we wil repeat all the steps described above. Note that there is no need to define again the functions created above because they are stored in current R session environment.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">setwd</span><span class="p">(</span><span class="s">&quot;/path/to/table&quot;</span><span class="p">)</span>
<span class="n">table.lit</span> <span class="o">=</span> <span class="nf">read.delim</span><span class="p">(</span><span class="s">&quot;taxa_abundance_bracken_names_normalized_literature.txt&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span> <span class="n">fill</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span> <span class="n">row.names</span><span class="o">=</span><span class="m">1</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s">&quot;\t&quot;</span><span class="p">)</span>
<span class="n">table.lit.final</span> <span class="o">=</span> <span class="n">table.lit</span><span class="p">[,</span><span class="o">-</span><span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="o">:</span><span class="m">3</span><span class="p">)]</span>
<span class="n">result.filter</span> <span class="o">=</span> <span class="nf">low.count.removal</span><span class="p">(</span><span class="nf">t</span><span class="p">(</span><span class="n">table.lit.final</span><span class="p">),</span> <span class="n">percent</span><span class="o">=</span><span class="m">0.02</span><span class="p">)</span>
<span class="n">table.lit.final.flt</span> <span class="o">=</span> <span class="n">result.filter</span><span class="o">$</span><span class="n">data.filter</span>
<span class="n">table.lit.final.flt.tss</span> <span class="o">=</span> <span class="nf">t</span><span class="p">(</span><span class="nf">apply</span><span class="p">(</span><span class="n">table.lit.final.flt</span><span class="p">,</span> <span class="m">1</span><span class="p">,</span> <span class="n">TSS.divide</span><span class="p">))</span>
</pre></div>
</div>
<p>Now we can merge the two tables in one, by merging them for the column containing the species names (this column is selected with <code class="docutils literal notranslate"><span class="pre">by=0</span></code>)</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">table.total</span> <span class="o">=</span> <span class="nf">merge</span><span class="p">(</span><span class="nf">t</span><span class="p">(</span><span class="n">mytable.final.flt.tss</span><span class="p">),</span> <span class="nf">t</span><span class="p">(</span><span class="n">table.lit.final.flt.tss</span><span class="p">),</span> <span class="n">by</span><span class="o">=</span><span class="m">0</span><span class="p">,</span> <span class="n">all</span><span class="o">=</span><span class="kc">TRUE</span><span class="p">)</span>
</pre></div>
</div>
<p>The following commands are used to finalize the table:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">table.total</span><span class="p">[</span><span class="nf">is.na</span><span class="p">(</span><span class="n">table.total</span><span class="p">)]</span> <span class="o">&lt;-</span> <span class="m">0</span>          <span class="c1">#removes NA</span>
<span class="nf">row.names</span><span class="p">(</span><span class="n">table.total</span><span class="p">)</span> <span class="o">=</span> <span class="n">table.total</span><span class="p">[,</span><span class="m">1</span><span class="p">]</span>      <span class="c1">#copy the species names in 1st columns to row names</span>
<span class="n">table.total.final</span> <span class="o">=</span> <span class="nf">t</span><span class="p">(</span><span class="n">table.total</span><span class="p">[,</span><span class="m">-1</span><span class="p">])</span>       <span class="c1">#delete the first column with the species names (now reported as row names)</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>You can merge datasets only if they were generated with the same taxonomy database (here the Minikraken 8Gb database). If not, you will have to run all the samples from the literature with the same database that you used to analyse your samples.</p>
</div>
</section>
<section id="upgma">
<h4><span class="section-number">10.3.1.2. </span>UPGMA<a class="headerlink" href="#upgma" title="Permalink to this headline"></a></h4>
<p>Once generated the final including both datasets (dental calculus and other microbiomes), we run an UPGMA cluster analysis. We must first install the <code class="docutils literal notranslate"><span class="pre">vegan</span></code> and <code class="docutils literal notranslate"><span class="pre">ape</span></code> package in R.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;vegan&quot;</span><span class="p">)</span>             <span class="c1">#do it only if the package is not installed yet</span>
<span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;ape&quot;</span><span class="p">)</span>               <span class="c1">#do it only if the package is not installed yet</span>
<span class="nf">library</span><span class="p">(</span><span class="n">vegan</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ape</span><span class="p">)</span>
</pre></div>
</div>
<p>Then we use vegan to calculate the <strong>Bray-Curtis</strong> distances, and run the cluster analysis.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">bray_dist</span> <span class="o">=</span> <span class="nf">vegdist</span><span class="p">(</span><span class="n">table.total.final</span><span class="p">,</span> <span class="n">method</span> <span class="o">=</span> <span class="s">&quot;bray&quot;</span><span class="p">,</span> <span class="n">binary</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> <span class="n">diag</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> <span class="n">upper</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> <span class="n">na.rm</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">)</span>
<span class="n">bray_dist.clust</span> <span class="o">=</span> <span class="nf">hclust</span><span class="p">(</span><span class="n">bray_dist</span><span class="p">,</span> <span class="n">method</span><span class="o">=</span><span class="s">&quot;average&quot;</span><span class="p">,</span> <span class="n">members</span> <span class="o">=</span> <span class="kc">NULL</span><span class="p">)</span>
</pre></div>
</div>
<p>Finally, we plot the dendrogram:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">plot</span><span class="p">(</span><span class="nf">as.phylo</span><span class="p">(</span><span class="n">bray_dist.clust</span><span class="p">),</span> <span class="n">type</span> <span class="o">=</span> <span class="s">&quot;unrooted&quot;</span><span class="p">,</span> <span class="n">cex</span> <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span> <span class="n">lab4ut</span><span class="o">=</span><span class="s">&quot;axial&quot;</span><span class="p">,</span> <span class="n">no.margin</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span> <span class="n">show.tip.label</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span> <span class="n">label.offset</span><span class="o">=</span><span class="m">0.02</span><span class="p">,</span> <span class="n">edge.color</span> <span class="o">=</span> <span class="s">&quot;gray&quot;</span><span class="p">,</span> <span class="n">edge.width</span> <span class="o">=</span> <span class="m">1</span><span class="p">,</span> <span class="n">edge.lty</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
</pre></div>
</div>
<p>To visualize better our samples in the following charts, we can define metadata as vectors. We assign group labels on each sample, creating a vector of labels that corresponds to the samples of the dataset that we are analysing.
For example, for the literature samples, we generate the following vector of metadata describing the kind of sample.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">labels_lit</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;Ancient calculus&quot;</span><span class="p">,</span><span class="s">&quot;Ancient tooth&quot;</span><span class="p">,</span><span class="s">&quot;Ancient calculus&quot;</span><span class="p">,</span><span class="s">&quot;Ancient tooth&quot;</span><span class="p">,</span>
                                      <span class="s">&quot;Soil&quot;</span><span class="p">,</span><span class="s">&quot;Soil&quot;</span><span class="p">,</span><span class="s">&quot;Soil&quot;</span><span class="p">,</span><span class="s">&quot;Soil&quot;</span><span class="p">,</span><span class="s">&quot;Soil&quot;</span><span class="p">,</span><span class="s">&quot;Soil&quot;</span><span class="p">,</span><span class="s">&quot;Soil&quot;</span><span class="p">,</span>
                                      <span class="s">&quot;Ancient calculus&quot;</span><span class="p">,</span><span class="s">&quot;Ancient tooth&quot;</span><span class="p">,</span><span class="s">&quot;Ancient calculus&quot;</span><span class="p">,</span><span class="s">&quot;Ancient tooth&quot;</span><span class="p">,</span><span class="s">&quot;Ancient calculus&quot;</span><span class="p">,</span><span class="s">&quot;Ancient tooth&quot;</span><span class="p">,</span><span class="s">&quot;Ancient calculus&quot;</span><span class="p">,</span><span class="s">&quot;Ancient tooth&quot;</span><span class="p">,</span><span class="s">&quot;Ancient calculus&quot;</span><span class="p">,</span><span class="s">&quot;Ancient tooth&quot;</span><span class="p">,</span><span class="s">&quot;Ancient calculus&quot;</span><span class="p">,</span><span class="s">&quot;Ancient tooth&quot;</span><span class="p">,</span>
                                      <span class="s">&quot;Plaque&quot;</span><span class="p">,</span><span class="s">&quot;Plaque&quot;</span><span class="p">,</span><span class="s">&quot;Plaque&quot;</span><span class="p">,</span><span class="s">&quot;Plaque&quot;</span><span class="p">,</span><span class="s">&quot;Plaque&quot;</span><span class="p">,</span><span class="s">&quot;Plaque&quot;</span><span class="p">,</span><span class="s">&quot;Plaque&quot;</span><span class="p">,</span><span class="s">&quot;Plaque&quot;</span><span class="p">,</span><span class="s">&quot;Plaque&quot;</span><span class="p">,</span><span class="s">&quot;Plaque&quot;</span><span class="p">,</span>
                                      <span class="s">&quot;Plaque&quot;</span><span class="p">,</span><span class="s">&quot;Plaque&quot;</span><span class="p">,</span><span class="s">&quot;Plaque&quot;</span><span class="p">,</span><span class="s">&quot;Plaque&quot;</span><span class="p">,</span><span class="s">&quot;Plaque&quot;</span><span class="p">,</span><span class="s">&quot;Plaque&quot;</span><span class="p">,</span><span class="s">&quot;Plaque&quot;</span><span class="p">,</span><span class="s">&quot;Plaque&quot;</span><span class="p">,</span><span class="s">&quot;Plaque&quot;</span><span class="p">,</span><span class="s">&quot;Plaque&quot;</span><span class="p">,</span>
                                      <span class="s">&quot;Plaque&quot;</span><span class="p">,</span><span class="s">&quot;Plaque&quot;</span><span class="p">,</span><span class="s">&quot;Plaque&quot;</span><span class="p">,</span><span class="s">&quot;Plaque&quot;</span><span class="p">,</span><span class="s">&quot;Plaque&quot;</span><span class="p">,</span>
                                      <span class="s">&quot;Skin&quot;</span><span class="p">,</span><span class="s">&quot;Skin&quot;</span><span class="p">,</span><span class="s">&quot;Skin&quot;</span><span class="p">,</span><span class="s">&quot;Skin&quot;</span><span class="p">,</span>
                                      <span class="s">&quot;Gut&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">,</span>
                                      <span class="s">&quot;Gut&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">,</span>
                                      <span class="s">&quot;Gut&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">,</span><span class="s">&quot;Gut&quot;</span><span class="p">,</span>
                                      <span class="s">&quot;Skin&quot;</span><span class="p">,</span><span class="s">&quot;Skin&quot;</span><span class="p">,</span><span class="s">&quot;Skin&quot;</span><span class="p">,</span><span class="s">&quot;Skin&quot;</span><span class="p">,</span><span class="s">&quot;Skin&quot;</span><span class="p">,</span>
                                      <span class="s">&quot;Plaque&quot;</span><span class="p">,</span><span class="s">&quot;Plaque&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Create a vector with labels corresponding to the samples that you analysed. Always make sure to follow the order of the samples if you use more than one label.
For example, the metadata of a table from dataset of Velsko containing ancient samples in the first five rows, and modern samples in the following five rows, will be represented by this vector:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">labels_Velsko</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="s">&quot;Velsko_ancient&quot;</span><span class="p">,</span><span class="s">&quot;Velsko_ancient&quot;</span><span class="p">,</span><span class="s">&quot;Velsko_ancient&quot;</span><span class="p">,</span><span class="s">&quot;Velsko_ancient&quot;</span><span class="p">,</span><span class="s">&quot;Velsko_ancient&quot;</span><span class="p">,</span>
                      <span class="s">&quot;Velsko_modern&quot;</span><span class="p">,</span><span class="s">&quot;Velsko_modern&quot;</span><span class="p">,</span><span class="s">&quot;Velsko_modern&quot;</span><span class="p">,</span><span class="s">&quot;Velsko_modern&quot;</span><span class="p">,</span><span class="s">&quot;Velsko_modern&quot;</span><span class="p">)</span>
</pre></div>
</div>
<p>Merge the labels, again paying attention to the order that you used to merge the tables (first your samples, then the literature dataset)</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">labels</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="n">labels_Velsko</span><span class="p">,</span><span class="n">labels_lit</span><span class="p">)</span>
</pre></div>
</div>
<p>To have a better look at the correspondence of data we can create a dataframe:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">table.total.final.df</span> <span class="o">=</span> <span class="nf">as.data.frame</span><span class="p">(</span><span class="n">table.total.final</span><span class="p">)</span>
<span class="n">table.total.final.df</span><span class="o">$</span><span class="n">group</span> <span class="o">=</span> <span class="n">labels</span>
</pre></div>
</div>
<p>We assign colors to each label. Note that the colors are assigned alphabetically based on the labels that you used.</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">coul</span><span class="o">=</span><span class="nf">c</span><span class="p">(</span><span class="s">&quot;#E41A1C&quot;</span><span class="p">,</span>             <span class="c1">#Ancient calculus</span>
      <span class="s">&quot;#419681&quot;</span><span class="p">,</span>              <span class="c1">#Ancient tooth</span>
      <span class="s">&quot;#4DAF4A&quot;</span><span class="p">,</span>              <span class="c1">#Gut</span>
      <span class="s">&quot;lightgray&quot;</span><span class="p">,</span>            <span class="c1">#Plaque</span>
      <span class="s">&quot;#984EA3&quot;</span><span class="p">,</span>              <span class="c1">#Skin</span>
      <span class="s">&quot;#FF7F00&quot;</span><span class="p">,</span>              <span class="c1">#Soil</span>
      <span class="s">&quot;goldenrod&quot;</span><span class="p">,</span>            <span class="c1">#Velsko-ancient</span>
      <span class="s">&quot;#994C00&quot;</span><span class="p">)</span>              <span class="c1">#Velsko-modern</span>
</pre></div>
</div>
<p>And finally, we plot the dendrogram by customizing the tips with the color-coded labels:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">plot</span><span class="p">(</span><span class="nf">as.phylo</span><span class="p">(</span><span class="n">bray_dist.clust</span><span class="p">),</span> <span class="n">type</span> <span class="o">=</span> <span class="s">&quot;unrooted&quot;</span><span class="p">,</span> <span class="n">cex</span> <span class="o">=</span> <span class="m">0.5</span><span class="p">,</span> <span class="n">lab4ut</span><span class="o">=</span><span class="s">&quot;axial&quot;</span><span class="p">,</span> <span class="n">no.margin</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span> <span class="n">show.tip.label</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span> <span class="n">label.offset</span><span class="o">=</span><span class="m">0.02</span><span class="p">,</span> <span class="n">edge.color</span> <span class="o">=</span> <span class="s">&quot;gray&quot;</span><span class="p">,</span> <span class="n">edge.width</span> <span class="o">=</span> <span class="m">1</span><span class="p">,</span> <span class="n">edge.lty</span> <span class="o">=</span> <span class="m">1</span><span class="p">)</span>
<span class="nf">tiplabels</span><span class="p">(</span><span class="n">pch</span><span class="o">=</span><span class="m">19</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="n">coul</span><span class="p">[</span><span class="nf">factor</span><span class="p">(</span><span class="n">labels</span><span class="p">)],</span> <span class="n">bg</span> <span class="o">=</span> <span class="n">coul</span><span class="p">[</span><span class="nf">factor</span><span class="p">(</span><span class="n">labels</span><span class="p">)],</span> <span class="n">cex</span><span class="o">=</span><span class="m">1</span><span class="p">,</span> <span class="n">lwd</span><span class="o">=</span><span class="m">1</span><span class="p">)</span>
</pre></div>
</div>
<p>And we can add a legend:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">legend</span><span class="p">(</span><span class="s">&quot;topleft&quot;</span><span class="p">,</span> <span class="n">legend</span> <span class="o">=</span> <span class="nf">sort</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="n">labels</span><span class="p">)),</span> <span class="n">bty</span> <span class="o">=</span> <span class="s">&quot;n&quot;</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="n">coul</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">19</span><span class="p">,</span> <span class="n">pt.cex</span><span class="o">=</span><span class="m">1</span><span class="p">,</span> <span class="n">cex</span><span class="o">=</span><span class="m">0.6</span><span class="p">,</span> <span class="n">pt.lwd</span><span class="o">=</span><span class="m">1</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="pca">
<h4><span class="section-number">10.3.1.3. </span>PCA<a class="headerlink" href="#pca" title="Permalink to this headline"></a></h4>
<p>Microbiome data are considered as compositional (see <a class="reference external" href="https://www.frontiersin.org/articles/10.3389/fmicb.2017.02224/full">Gloor et al. 2017</a>), for this reason we will apply the Centered-Log-Ratio (CLR) normalization with the library <code class="docutils literal notranslate"><span class="pre">composition</span></code> and the package <code class="docutils literal notranslate"><span class="pre">mixOmics</span></code>. We will use also the library <code class="docutils literal notranslate"><span class="pre">ape</span></code>.
To install mixOmics follow the <a class="reference external" href="http://mixomics.org/install/">instructions</a>.</p>
<blockquote>
<div></div></blockquote>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;compositions&quot;</span><span class="p">)</span>              <span class="c1">#do it only if the package is not installed yet</span>
<span class="nf">install.packages</span><span class="p">(</span><span class="s">&quot;shape&quot;</span><span class="p">)</span>             <span class="c1">#do it only if the package is not installed yet</span>
<span class="nf">library</span><span class="p">(</span><span class="n">compositions</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">ape</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">mixOmics</span><span class="p">)</span>
<span class="nf">library</span><span class="p">(</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
<p>Compositional data methods depend on logarithms that do not compute for <a class="reference external" href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6755255/">zeros</a>, for this reason they have to be removed from the dataset. We will account for that by using an offset, e.g. by adding one count (+1) to the entire abundance table. Pay attention to the order of magnitude of your data and choose the offset accordingly (the offset must represent a very small fraction of the smallest values observed in the table). To do that, we will convert the table in counts per milion (<a href="#id1"><span class="problematic" id="id2">*</span></a>1000000), and use the offset +1.</p>
<blockquote>
<div></div></blockquote>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">table.total.final.off</span><span class="o">=</span><span class="p">(</span><span class="n">table.total.final</span><span class="o">*</span><span class="m">1000000</span><span class="p">)</span><span class="m">+1</span>
<span class="nf">sum</span><span class="p">(</span><span class="nf">which</span><span class="p">(</span><span class="n">table.total.final.off</span> <span class="o">==</span> <span class="m">0</span><span class="p">))</span>                <span class="c1"># to check if any zero is left</span>
</pre></div>
</div>
<p>We can now run the PCA with <code class="docutils literal notranslate"><span class="pre">tune.pca</span></code> from mixOmics, using the CLR normalization and setting 10 compoments. You can find more detailed instructions in the mixOmics <a class="reference external" href="http://mixomics.org/methods/pca/">website</a>.</p>
<blockquote>
<div></div></blockquote>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">tune.pca</span><span class="p">(</span><span class="n">table.total.final.off</span><span class="p">,</span> <span class="n">ncomp</span> <span class="o">=</span> <span class="m">10</span><span class="p">,</span> <span class="n">center</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">scale</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> <span class="n">logratio</span> <span class="o">=</span> <span class="s">&#39;CLR&#39;</span><span class="p">)</span>
<span class="n">pca.res</span> <span class="o">&lt;-</span> <span class="nf">pca</span><span class="p">(</span><span class="n">table.total.final.off</span><span class="p">,</span> <span class="n">ncomp</span> <span class="o">=</span> <span class="m">10</span><span class="p">,</span> <span class="n">center</span> <span class="o">=</span> <span class="kc">TRUE</span><span class="p">,</span> <span class="n">scale</span> <span class="o">=</span> <span class="kc">FALSE</span><span class="p">,</span> <span class="n">logratio</span> <span class="o">=</span> <span class="s">&#39;CLR&#39;</span><span class="p">)</span>
</pre></div>
</div>
<p>You can investigate and plot the loadings of the PCA with the following commands:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="n">pca.res</span><span class="o">$</span><span class="n">sdev</span><span class="o">^</span><span class="m">2</span>
<span class="n">pca.res</span><span class="o">$</span><span class="n">rotation</span>
<span class="n">X</span> <span class="o">&lt;-</span> <span class="n">pca.res</span><span class="o">$</span><span class="n">rotation</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">X</span><span class="p">,</span> <span class="n">asp</span><span class="o">=</span><span class="m">1</span><span class="p">,</span> <span class="n">type</span><span class="o">=</span><span class="s">&quot;n&quot;</span><span class="p">)</span>              <span class="c1"># To plot the frame</span>
<span class="nf">abline</span><span class="p">(</span><span class="n">v</span><span class="o">=</span><span class="m">0</span><span class="p">,</span> <span class="n">lty</span><span class="o">=</span><span class="m">3</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s">&quot;grey&quot;</span><span class="p">)</span>
<span class="nf">abline</span><span class="p">(</span><span class="n">h</span><span class="o">=</span><span class="m">0</span><span class="p">,</span> <span class="n">lty</span><span class="o">=</span><span class="m">3</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s">&quot;grey&quot;</span><span class="p">)</span>
<span class="nf">Arrows</span><span class="p">(</span><span class="m">0</span><span class="p">,</span> <span class="m">0</span><span class="p">,</span> <span class="n">X</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span> <span class="n">X</span><span class="p">[,</span><span class="m">2</span><span class="p">],</span> <span class="n">col</span><span class="o">=</span><span class="s">&quot;red&quot;</span><span class="p">,</span> <span class="n">code</span><span class="o">=</span><span class="m">2</span><span class="p">,</span> <span class="n">arr.length</span> <span class="o">=</span> <span class="m">0.15</span><span class="p">,</span> <span class="n">arr.width</span> <span class="o">=</span> <span class="m">0.1</span><span class="p">,</span> <span class="n">arr.type</span><span class="o">=</span><span class="s">&quot;triangle&quot;</span><span class="p">)</span>   <span class="c1"># To plot the arrows: see ?arrows for the syntax</span>
<span class="nf">text</span><span class="p">(</span><span class="m">1.06</span><span class="o">*</span><span class="n">X</span><span class="p">,</span> <span class="nf">rownames</span><span class="p">(</span><span class="n">X</span><span class="p">),</span> <span class="n">col</span><span class="o">=</span><span class="s">&quot;red&quot;</span><span class="p">,</span> <span class="n">xpd</span><span class="o">=</span><span class="bp">T</span><span class="p">,</span> <span class="n">cex</span><span class="o">=</span><span class="m">0.5</span><span class="p">)</span>          <span class="c1"># To label the arrows</span>
</pre></div>
</div>
<p>To plot the PCA in a two-dimensional plot use the following commands:</p>
<div class="highlight-r notranslate"><div class="highlight"><pre><span></span><span class="nf">par</span><span class="p">(</span><span class="n">mfrow</span> <span class="o">=</span> <span class="nf">c</span><span class="p">(</span><span class="m">1</span><span class="p">,</span> <span class="m">1</span><span class="p">))</span>          <span class="c1">#set the window for the chart (single chart, so 1 column and 1 row)</span>
<span class="nf">plot</span><span class="p">(</span><span class="n">pca.res</span><span class="o">$</span><span class="n">x</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span> <span class="n">pca.res</span><span class="o">$</span><span class="n">x</span><span class="p">[,</span><span class="m">2</span><span class="p">],</span> <span class="n">type</span><span class="o">=</span><span class="s">&quot;n&quot;</span><span class="p">,</span> <span class="n">main</span><span class="o">=</span><span class="s">&quot;PCA&quot;</span><span class="p">,</span> <span class="n">cex.axis</span><span class="o">=</span><span class="m">0.75</span><span class="p">,</span> <span class="n">cex.lab</span><span class="o">=</span><span class="m">0.75</span><span class="p">,</span> <span class="n">xlab</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">,</span> <span class="n">ylab</span><span class="o">=</span><span class="s">&quot;&quot;</span><span class="p">)</span>
<span class="nf">abline</span><span class="p">(</span><span class="n">v</span><span class="o">=</span><span class="m">0</span><span class="p">,</span> <span class="n">lty</span><span class="o">=</span><span class="m">3</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s">&quot;grey&quot;</span><span class="p">)</span>
<span class="nf">abline</span><span class="p">(</span><span class="n">h</span><span class="o">=</span><span class="m">0</span><span class="p">,</span> <span class="n">lty</span><span class="o">=</span><span class="m">3</span><span class="p">,</span> <span class="n">col</span><span class="o">=</span><span class="s">&quot;grey&quot;</span><span class="p">)</span>
<span class="nf">points</span><span class="p">(</span><span class="n">pca.res</span><span class="o">$</span><span class="n">x</span><span class="p">[,</span><span class="m">1</span><span class="p">],</span> <span class="n">pca.res</span><span class="o">$</span><span class="n">x</span><span class="p">[,</span><span class="m">2</span><span class="p">],</span> <span class="n">cex</span><span class="o">=</span><span class="m">1.2</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="n">coul</span><span class="p">[</span><span class="nf">factor</span><span class="p">(</span><span class="n">labels</span><span class="p">)],</span> <span class="n">bg</span> <span class="o">=</span> <span class="n">coul</span><span class="p">[</span><span class="nf">factor</span><span class="p">(</span><span class="n">labels</span><span class="p">)],</span> <span class="n">pch</span><span class="o">=</span><span class="m">19</span><span class="p">,</span> <span class="n">lwd</span><span class="o">=</span><span class="m">0.7</span><span class="p">)</span>
<span class="nf">legend</span><span class="p">(</span><span class="s">&quot;topleft&quot;</span><span class="p">,</span> <span class="n">legend</span> <span class="o">=</span> <span class="nf">sort</span><span class="p">(</span><span class="nf">unique</span><span class="p">(</span><span class="n">labels</span><span class="p">)),</span> <span class="n">bty</span> <span class="o">=</span> <span class="s">&quot;n&quot;</span><span class="p">,</span> <span class="n">col</span> <span class="o">=</span> <span class="n">coul</span><span class="p">,</span> <span class="n">pch</span> <span class="o">=</span> <span class="m">19</span><span class="p">,</span> <span class="n">pt.cex</span><span class="o">=</span><span class="m">1</span><span class="p">,</span> <span class="n">cex</span><span class="o">=</span><span class="m">0.6</span><span class="p">,</span> <span class="n">pt.lwd</span><span class="o">=</span><span class="m">1</span><span class="p">)</span>
</pre></div>
</div>
</section>
</section>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="7_Filtering_SNPs.html" class="btn btn-neutral float-left" title="9. Filtering, annotating and combining SNPs" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="8_HandsOn.html" class="btn btn-neutral float-right" title="11. Hands-on" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2021, Claudio Ottoni.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>
    <!-- Theme Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=UA-XXXXXXX-1"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());

      gtag('config', 'UA-XXXXXXX-1', {
          'anonymize_ip': false,
      });
    </script> 

</body>
</html>